---
title: "integrated"
output: html_document
date: "2024-06-22"
---

0. Integrate, regress out cell cycle, match TCR, plot XRTCCs, phenotype clusters
```{r}
# install umap-learn within R
# Check if umap-learn is available within the virtual environment you're using; If it returns FALSE, then umap-learn is not available in the specified Python environment.

Sys.setenv(RETICULATE_PYTHON='/Users/nathansuek/miniconda3/envs/r-reticulate/bin/python')
library(reticulate)
use_condaenv("/Users/nathansuek/miniconda3/envs/r-reticulate/bin/python", required = TRUE)
py_config()
R.Version()$arch
Sys.info()["machine"]
# # If not installed, can install umap-learn from within R
# library(reticulate)
# use_python("/Users/nathansuek/.pyenv/versions/3.10.13/bin/python", required = TRUE)
# py_install("umap-learn")
# 
# # Check if it's availble int h
# library(reticulate)
# use_virtualenv("~/.virtualenvs/r-reticulate", required = TRUE)
# py_module_available("umap")

# /Users/nathansuek/.pyenv/versions/3.10.13/lib/python3.10/site-packages/umap
```

```{r}
#########Integration all data to analyze them together#######################
library(Seurat)
library(tidyverse)
library(SeuratObject)
library(cluster)
library(ggplot2)
library(EnhancedVolcano)

###########read the data########################
path = 'Z:/CCTI_SYKES/EXP FF7- XKLT-001 Exp/single cell RNA analysis XKLT001 Analysed data/'
#path = '/Volumes/CCTI_LABS/CCTI_SYKES/EXP FF7- XKLT-001 Exp/single cell RNA analysis XKLT001 Analysed data/'
POD14k = Read10X(data.dir = paste0(path, 'POD 14 Kidney/KB03_cellranger_count_outs/KB03_cellranger_count_outs/filtered_feature_bc_matrix'))
POD14L = Read10X(data.dir = paste0(path, 'POD 14 Lymphocele/KB02_cellranger_count_outs (1)/KB02_cellranger_count_outs/filtered_feature_bc_matrix/' ))
POD28 = Read10X(data.dir = paste0(path, 'POD 28/filtered_feature_bc_matrix/'))
POD33 = Read10X(data.dir = paste0(path, 'POD 33/KB05_cellranger_count_outs (1)/KB05_cellranger_count_outs/filtered_feature_bc_matrix/' ))
POD61 = Read10X(data.dir = paste0(path, 'POD 61/XKLT001-POD61GE/filtered_feature_bc_matrix/' ))

###############creat seurat objects################
POD14k = CreateSeuratObject(counts = POD14k, min.features = 500, min.cells = 50)
POD14L = CreateSeuratObject(counts = POD14L, min.features = 500, min.cells = 50)
POD28 = CreateSeuratObject(counts = POD28, min.features = 500, min.cells = 50)
POD33 = CreateSeuratObject(counts = POD33, min.features = 500, min.cells = 50)
POD61 = CreateSeuratObject(counts = POD61, min.features = 300, min.cells = 50)

######## clean mitochondrial DNA ###########
##POD14K###
POD14k = PercentageFeatureSet(POD14k, pattern = '^MT-', col.name = 'percent.mt')
POD14k = subset(POD14k, subset = percent.mt<15 & nCount_RNA>1000 & nCount_RNA <25000)

######POD14L######
POD14L = PercentageFeatureSet(POD14L, pattern = '^MT-', col.name = 'percent.mt')
POD14L = subset(POD14L, subset = percent.mt<15 & nCount_RNA>1000 & nCount_RNA <25000)

######POD28######
POD28 = PercentageFeatureSet(POD28, pattern = '^MT-', col.name = 'percent.mt')
POD28 = subset(POD28, subset = percent.mt<15 & nCount_RNA>1000 & nCount_RNA <25000)

#########POD33###########
POD33 = PercentageFeatureSet(POD33, pattern = '^MT-', col.name = 'percent.mt')
POD33 = subset(POD33, subset = percent.mt<15 & nCount_RNA>1000 & nCount_RNA <30000)

#########POD61###########
POD61 = PercentageFeatureSet(POD61, pattern = '^MT-', col.name = 'percent.mt')
POD61 = subset(POD61, subset = percent.mt<15 & nCount_RNA>1000 & nCount_RNA <30000)

############Normalized the data##############
POD14k = SCTransform(POD14k, return.only.var.genes = F, verbose = T, conserve.memory = F)
POD14L = SCTransform(POD14L, return.only.var.genes = F, verbose = T, conserve.memory = F)
POD28 = SCTransform(POD28, return.only.var.genes = F, verbose = T, conserve.memory = F)
POD33 = SCTransform(POD33, return.only.var.genes = F, verbose = T, conserve.memory = F)
POD61 = SCTransform(POD61, return.only.var.genes = F, verbose = T, conserve.memory = F)

############ Add metadata for split the integrated data in next step ############
POD14k@meta.data$POD = 'POD 14K'
POD14L@meta.data$POD = 'POD 14L'
POD28@meta.data$POD = 'POD 28'
POD33@meta.data$POD = 'POD 33'
POD61@meta.data$POD = 'POD 61'

POD14k@meta.data$rejection = 'non-rejecting'
POD14L@meta.data$rejection = 'non-rejecting'
POD28@meta.data$rejection = 'non-rejecting'
POD33@meta.data$rejection  = 'rejecting'
POD61@meta.data$rejection  = 'rejecting'

#######integrate the data#######
list_cells =list(POD14k,POD14L, POD28, POD33, POD61)
features = SelectIntegrationFeatures(object.list = list_cells, nfeatures = 25000)
list_cells = PrepSCTIntegration(object.list = list_cells, anchor.features = features, verbose = T)
anchors = FindIntegrationAnchors(object.list = list_cells, normalization.method = 'SCT', anchor.features = features, verbose = T, reference = 1)
Integrated_Data = IntegrateData(anchorset = anchors, normalization.method = 'SCT', verbose = T, k.weight = 31)


# Regress out cell cycle genes
DefaultAssay(Integrated_Data) = 'RNA'
Integrated_Data = NormalizeData(Integrated_Data)
Integrated_Data = ScaleData(Integrated_Data)
Integrated_Data = JoinLayers(Integrated_Data)
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
Integrated_Data <- CellCycleScoring(Integrated_Data, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
Integrated_Data <- RunPCA(Integrated_Data, features = c(s.genes, g2m.genes))
DimPlot(Integrated_Data)


Integrated_Data <- ScaleData(Integrated_Data, vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(Integrated_Data))
Integrated_Data <- RunPCA(Integrated_Data, features = c(s.genes, g2m.genes))
DimPlot(Integrated_Data)



#######################################################################################################
####clustring###########
Integrated_Data <- FindVariableFeatures(Integrated_Data, selection.method = "vst", nfeatures = 2000)
Integrated_Data = RunPCA(Integrated_Data, features = VariableFeatures(object = Integrated_Data))
Integrated_Data = RunUMAP(Integrated_Data, dims = 1:30, verbose = FALSE, umap.method = 'umap-learn', metric = 'correlation')
Integrated_Data = FindNeighbors(Integrated_Data, dims = 1:30, verbose = F)
Integrated_Data = FindClusters(Integrated_Data, resolution = seq(0.01,2, by= 0.1), verbose = F, algorithm = 1)
DimPlot(Integrated_Data, label = T, group.by = 'RNA_snn_res.1.21')

dat = Integrated_Data








# Link TCR data
path = 'Z:/CCTI_SYKES/EXP FF7- XKLT-001 Exp/single cell RNA analysis XKLT001 Analysed data/'
TCR_POD14L = read.csv(paste0(path, 'filtered_contig_annotations_POD14L.csv'))
TCR_POD28 = read.csv(paste0(path, 'filtered_contig_annotations_POD28.csv'))
TCR_POD33 = read.csv(paste0(path, 'filtered_contig_annotations_POD33.csv'))
TCR_POD61 = read.csv(paste0(path, 'filtered_contig_annotations_POD61.csv'))

TCR_POD14L$pt = 'POD14L'
TCR_POD28$pt = 'POD28'
TCR_POD33$pt = 'POD33'
TCR_POD61$pt = 'POD61'

# Check if there are identical barcodes between the different samples (this would throw off matching based on cell barcode in next step)
temp = sum(TCR_POD14L$barcode %in% TCR_POD28$barcode) #0
temp = sum(TCR_POD14L$barcode %in% TCR_POD33$barcode) #0
temp = sum(TCR_POD14L$barcode %in% TCR_POD61$barcode) #0
temp = sum(TCR_POD28$barcode %in% TCR_POD33$barcode) #0
temp = sum(TCR_POD14L$barcode %in% TCR_POD61$barcode) #0
temp = sum(TCR_POD33$barcode %in% TCR_POD61$barcode) #0

#tcr_seq = rbind(TCR_POD14L, TCR_POD28, TCR_POD33, TCR_POD61)
tcr_seq = rbind(TCR_POD14L, TCR_POD28, TCR_POD33, TCR_POD61)

dat@meta.data$barcodes = rownames(dat@meta.data)
dat@meta.data$barcodes <- gsub("_\\d$", "", dat@meta.data$barcodes)

# Assign TCRaa CDR3 sequence in metadata
dat@meta.data$TCRa_1 = 0
dat@meta.data$TCRa_2 = 0
dat@meta.data$TCRa_3 = 0
dat@meta.data$TCRb_1 = 0
dat@meta.data$TCRb_2 = 0
dat@meta.data$TCRb_3 = 0
l1 = nrow(dat@meta.data)

for(i in 1:l1){
  ind = which(tcr_seq == dat@meta.data$barcodes[i])
  
  ind_2 = which(tcr_seq[ind,]$chain == 'TRA')
  l2 = length(ind_2)
  
  for(j in 1:l2){
    if(l2 == 1)
      dat@meta.data$TCRa_1[i] = tcr_seq[ind[ind_2[1]],]$cdr3
    if(l2 == 2){
      dat@meta.data$TCRa_1[i] = tcr_seq[ind[ind_2[1]],]$cdr3
      dat@meta.data$TCRa_2[i] = tcr_seq[ind[ind_2[2]],]$cdr3
    }
    if(l2 == 3){
      dat@meta.data$TCRa_1[i] = tcr_seq[ind[ind_2[1]],]$cdr3
      dat@meta.data$TCRa_2[i] = tcr_seq[ind[ind_2[2]],]$cdr3
      dat@meta.data$TCRa_3[i] = tcr_seq[ind[ind_2[3]],]$cdr3
    }
  }
  
  ind_3 = which(tcr_seq[ind,]$chain == 'TRB')
  l3 = length(ind_3)
  for(j in 1:l3){
    if(l3 == 1)
      dat@meta.data$TCRb_1[i] = tcr_seq[ind[ind_3[1]],]$cdr3
    if(l3 == 2){
      dat@meta.data$TCRb_1[i] = tcr_seq[ind[ind_3[1]],]$cdr3
      dat@meta.data$TCRb_2[i] = tcr_seq[ind[ind_3[2]],]$cdr3
    }
    if(l3 == 3){
      dat@meta.data$TCRb_1[i] = tcr_seq[ind[ind_3[1]],]$cdr3
      dat@meta.data$TCRb_2[i] = tcr_seq[ind[ind_3[2]],]$cdr3
      dat@meta.data$TCRb_3[i] = tcr_seq[ind[ind_3[3]],]$cdr3
    }
  }
}





# Identify XRTCCs
path = 'Z:/CCTI_SYKES/EXP FF7- XKLT-001 Exp/Bulk TCRb sequincing Adaptive Data XKLT001/All MLRs, Official Analysis/'
DRTCC_CD8 = read.csv(paste0(path, 'MasterCD8DRTCCs_AgainstPreTx.csv'))
DRTCC_CD4 = read.csv(paste0(path, 'MasterCD4DRTCCs_AgainstPreTx.csv'))
nonDRTCC = read.csv(paste0(path, 'PreTxNonDRTCCsConservative.csv'))

meta = dat@meta.data

# DRTCC
ind = which(meta$TCRb_1 %in% DRTCC_CD8$Amino.Acid | meta$TCRb_2 %in% DRTCC_CD8$Amino.Acid | meta$TCRb_3 %in% DRTCC_CD8$Amino.Acid) 
barcode = meta[ind,] %>% row.names()

ind = which(meta$TCRb_1 %in% DRTCC_CD4$Amino.Acid | meta$TCRb_2 %in% DRTCC_CD4$Amino.Acid | meta$TCRb_3 %in% DRTCC_CD4$Amino.Acid)
barcode2 = meta[ind,] %>% row.names()

highlight = c(barcode, barcode2)

# Non DRTCC
ind = which(meta$TCRb_1 %in% nonDRTCC$Amino.Acid | meta$TCRb_2 %in% nonDRTCC$Amino.Acid | meta$TCRb_3 %in% nonDRTCC$Amino.Acid)
barcode3 = meta[ind,] %>% row.names()

highlight2 = c(barcode3)

DimPlot(dat, split.by = 'POD', cells.highlight = highlight) # Plot XRTCCs

dat$XRTCC = ''
ind = which(row.names(dat@meta.data) %in% highlight)
dat$XRTCC[ind] = 'XRTCC'

ind = which(row.names(dat@meta.data) %in% highlight2)
dat$XRTCC[ind] = 'nonXRTCC'

DimPlot(dat, group.by = 'XRTCC', cols = c('lightgray', 'blue', 'red'), pt.size = 1.5, order = c('XRTCC', 'nonXRTCC', ''), split.by = 'XRTCC')
DimPlot(dat, group.by = 'XRTCC', cols = c('lightgray', 'blue', 'red'), pt.size = 1.5, order = c('XRTCC', 'nonXRTCC', ''), split.by = 'POD')


# Optimize cluster resolution
# Summary: decision is done manually based on (1) how well heatmaps separate clusters (2) if it's clear which clusters appear at POD33 (3) where are the XRTCCs
# (0) Pick cluster resolution
Idents(dat) = 'RNA_snn_res.0.41'

# (1) XRTCC by cluster proportion
df = prop.table(table(Idents(dat), dat$POD), margin = 2)
colors = brewer.pal(5, "Set1")
barplot(t(as.matrix(df)),
        beside = TRUE,
        col = colors,
        legend = colnames(df),
        xlab = "Cluster",
        ylab = "Proportion",
        main = "Cluster proportions by POD")

# (2) XRTCC by cluster proportion
df = table(Idents(dat), dat$XRTCC)[,2]
colors = c('firebrick')
barplot(t(as.matrix(df)),
        beside = TRUE,
        col = colors,
        legend = colnames(df),
        xlab = "Cluster",
        ylab = "Number of XRTCCs",
        main = "XRTCCs by cluster")

# Marker Plotting
DefaultAssay(dat) = 'RNA'
dat = NormalizeData(dat)
var.features = VariableFeatures(dat)
extra.features = c('CD3E', 'CD3D', 'CD3G', 'GATA3', 'IL13', 'IL5', 'IL4', 'IL17A', 'IL17RA', 'IL22', 'IL22RA1', 'IL23R', 'RORC', 'RORA', 'TCF7', 'TMIGD2', 'CD28', 'SELL', 'S1PR1', 'KLF2', 'CCR7', 'ZNF683', 'CXCR6', 'ITGA1', 'ITGAE', 'TRDV2', 'TRBC1', 'NCR1', 'CD34', 'CD16', 'CD19', 'HLA-C', 'HLA-B', 'HLA-A', 'SATB1', 'IL2', 'TGFB1', 'MAF', 'STAT5A', 'BATF', 'IL21R', 'STAT3', 'IL6', 'IL21', 'CXCL13', 'CXCR5')
dat = ScaleData(dat, assay = 'RNA', features = c(var.features, extra.features))
dat = JoinLayers(dat)

jianing_genes1 = c("CD79A", "CD74", 'CD19', 'MS4A1', 'CD14', 'CD16', 'ITGAM', 'ITGAX', "CD34","CD3E", "CD3D", "CD3G", "CD4", "CD8A", 'CD8B', 'CD4', 'NCAM1', 'NCR1', "TRBC1","TRDC", 'TRGV9', 'TRDV2', 'IGHM', "CD69", "ITGAE", "ITGA1", "CXCR6", "RUNX3", "PRDM1", "ZNF683", "CCR7", "KLF2", "S1PR1", "SELL", "MKI67", "CD28","TMIGD2", "TIGIT","TCF7", "RORA", "RORC", "AHR", "IL23R", "IL23A", "IL22RA1", "IL22", "IL17RA","IL17A","CCL20", "TNF", "IFNG", "IFNGR1","TBX21", "EOMES", "IL4", "IL5", "IL13", "GATA3")

jianing_genes2 = c("BCL6", "CXCR5", "PDCD1", "ICOS", "IL2RA", "CXCL13", "IL21", "IL6", "STAT3", "IRF4", "IL21R", "IL6R", "BATF", "STAT5A", "MAF", "GZMB", "GZMA","PRF1","GNLY", "KLRD1", "KLRC1", "KLRK1","FOXP3", "CTLA4","IL10","TGFB1", "IL2","NR4A1", "SATB1", "EGR1", "IL4I1", "NFATC1","JUN", "FOS","NFKBIA", "NFKBID","CD83", "PTPN22", "CXCR4", "BTG2", "HLA-A", "HLA-B", "HLA-C", "HLA-DRB1", "HLA-DPB1", "HLA-DQB1", "LYN", "CCL4")

DoHeatmap(dat, features = jianing_genes1, assay = 'RNA') + scale_fill_gradientn(colors = c("cornflowerblue","coral3"))
DoHeatmap(dat, features = jianing_genes2, assay = 'RNA') + scale_fill_gradientn(colors = c("cornflowerblue","coral3"))

markers = FindAllMarkers(dat)
top = markers %>% group_by(cluster) %>% top_n(5, wt = avg_log2FC)
DoHeatmap(dat, features = top$gene, assay = 'RNA') + scale_fill_gradientn(colors = c("cornflowerblue","coral3"))

markers_c3 = markers %>% filter(cluster == '3')
EnhancedVolcano(markers_c3, lab = markers_c3$gene, x = 'avg_log2FC', y = 'p_val_adj', drawConnectors = T, pCutoff = 10e-100, max.overlaps = 20) + ggtitle('Cluster 3')

markers_c6 = markers %>% filter(cluster == '6')
EnhancedVolcano(markers_c6, lab = markers_c6$gene, x = 'avg_log2FC', y = 'p_val_adj', drawConnectors = T, pCutoff = 10e-50, max.overlaps = 20) + ggtitle('Cluster 6')
```

Plot each TCR individually 
```{r}
dat$XRTCC_cdr3 = ''
ind = which(row.names(dat@meta.data) %in% highlight)
dat$XRTCC_cdr3[ind] = dat$TCRb_1[ind]

TCRs = dat@meta.data %>% filter(XRTCC == 'XRTCC')
TCRs = TCRs$TCRb_1
TCRs = unique(TCRs)

for(i in 1:length(TCRs)){
  ind = TCRs[i]
  ind = which(dat@meta.data$TCRb_1 == ind)
  cell = row.names(dat@meta.data[ind,])
  p = DimPlot(dat, split.by = 'POD', cells.highlight = cell) + ggtitle(paste0(TCRs[i]))
  plot(p)
}
```

Compare Clusters 3 vs 6
Figure S7
```{r}
markers = FindMarkers(dat, ident.1 = '3', ident.2 = '6')
markers$gene = row.names(markers)
labels = c('NCAM', 'NCR1', 'NCR3', 'TRDC', 'TRGC1', 'TRGV9', 'TRGV7', 'GZMB', 'GZMK', 'GZMM', 'GZMA', 'GZMH', 'KLRK1', 'CD8B', 'CD3', 'CD3D', 'CD3E', 'CD3G', 'TRBC2')
EnhancedVolcano(markers, lab = row.names(markers), x = 'avg_log2FC', y = 'p_val_adj', drawConnectors = T, pCutoff = 10e-5, max.overlaps = 30, selectLab = labels) + ggtitle('Cluster 3 vs 6')


markers = FindAllMarkers(dat)
all = markers %>% group_by(cluster) %>% top_n(70, wt = avg_log2FC)
all = all %>% filter(cluster == '3' | cluster == '6')
top = all %>% filter(cluster == '3') %>% top_n(20, wt = avg_log2FC)
bottom = all %>% filter(cluster == '6') %>% top_n(60, wt = avg_log2FC)

select = c(top$gene, bottom$gene)
sub = subset(dat, idents = c('3', '6'))
DoHeatmap(sub, features = select, group.colors = c("#CECBD0", "#BD98A2")) + scale_fill_gradientn(colors = c("cornflowerblue","coral3")) +  theme(text = element_text(size = 20))




# Fig S7
comparison_list <- list(c('3', '6')) # List of comparisons
genes <- c('GZMB', "TNF", "IFNG", "IFNGR1", "TBX21", "EOMES", "PDCD1", "ICOS", "GZMB", "GZMA", "GZMK", "PRF1", "GNLY", "KLRD1", "KLRC1", "KLRK1") # List of gene sets
plots_adjusted <- list() # Initialize a list to store plots


# Generate the violin plot for the current set of genes
for(i in seq_along(genes)){
  plots <- VlnPlot(sub, features = genes[i]) + NoLegend()
  y_limits <- ggplot_build(plots)$layout$panel_scales_y[[1]]$range$range
  
  # Adjust ylim and add non-overlapping p-value labels
  plots_adjusted[[i]] = plots + 
    stat_compare_means(comparisons = comparison_list, 
                                 label = "p.signif", 
                                 method = 'wilcox.test',
                                 step.increase = 0.2, # Prevents the p-value symbols from overlapping on graph
                                 tip.length = 0) + # Length of the downward tip of each line for each comparison on the graph
          ylim(y_limits[1], y_limits[2] + 0.5) + # Extends y-axis of each graph to accommodate the p-value signs
          theme(axis.title.x = element_blank(), # Hide x-axis title
                axis.title.y = element_blank()) # Hide y-axis title
}
 
# print each plot individually, copy paste into ppt
combined_plot <- plot_grid(plotlist = plots_adjusted, ncol = 5, nrow = 3)
combined_plot
```

Subcluster XRTCCs
```{r}
# Violint Plots: cluster 3 vs 6
Idents(dat) = 'RNA_snn_res.0.41'
sub = subset(dat, idents = (c('3', '6')))

Idents(sub) = 'RNA_snn_res.0.41'
VlnPlot(sub, features = c('GZMB', 'GZMA', 'GZMM', 'GZMH', 'GZMK', 'PRF1', 'GNLY', 'KLRK1', 'KLRD1'))

# Subcluster XRTCCs
Idents(dat) = 'XRTCC'
sub = subset(dat, idents = (c('XRTCC')))

sub@assays$SCT = NULL
sub@assays$integrated = NULL
sub@meta.data = sub@meta.data[-c(13:38)]

sub = SCTransform(sub, return.only.var.genes = F, verbose = T, conserve.memory = F)
sub = RunPCA(sub, features = VariableFeatures(object = sub))
sub = RunUMAP(sub, dims = 1:30, verbose = FALSE, umap.method = 'umap-learn', metric = 'correlation')
sub = FindNeighbors(sub, dims = 1:30, verbose = F)
sub = FindClusters(sub, resolution = seq(0.01,2, by= 0.1), verbose = F, algorithm = 1)
DimPlot(sub, label = T)
DimPlot(sub, label = T, cells.highlight = highlight)

Idents(sub) = 'SCT_snn_res.1.11'
markers = FindAllMarkers(sub)
top = markers %>% group_by(cluster) %>% top_n(10, wt = avg_log2FC)
DoHeatmap(sub, features = top$gene) + scale_fill_gradientn(colors = c("cornflowerblue","coral3"))
```

Subcluster cluster 3 + plot AB vs yd T cells
Figure 6B
```{r}
# Violint Plots: cluster 3 vs 6
Idents(dat) = 'RNA_snn_res.0.41'
sub = subset(dat, idents = (c('3')))

sub@assays$SCT = NULL
sub@assays$integrated = NULL
sub@meta.data = sub@meta.data[-c(13:38)]

sub = SCTransform(sub, return.only.var.genes = F, verbose = T, conserve.memory = F)
sub = RunPCA(sub, features = VariableFeatures(object = sub))
sub = RunUMAP(sub, dims = 1:30, verbose = FALSE, umap.method = 'umap-learn', metric = 'correlation')
sub = FindNeighbors(sub, dims = 1:30, verbose = F)
sub = FindClusters(sub, resolution = seq(0.01,2, by= 0.1), verbose = F, algorithm = 1)
DimPlot(sub, label = T)
DimPlot(sub, label = T, group.by = 'XRTCC')

saveRDS(sub, file = 'sub_c3.rds')



sub = readRDS('sub_c3.rds')

# (2) XRTCC by cluster proportion
df = table(Idents(sub), sub$XRTCC)[,2]
colors = c('firebrick')
barplot(t(as.matrix(df)),
        beside = TRUE,
        col = colors,
        legend = colnames(df),
        xlab = "Cluster",
        ylab = "Number of XRTCCs",
        main = "XRTCCs by cluster")


markers = FindAllMarkers(sub)
top = markers %>% group_by(cluster) %>% top_n(10, wt = avg_log2FC)
DoHeatmap(sub, features = top$gene) + scale_fill_gradientn(colors = c("cornflowerblue","coral3"))



# Define the genes of interest
genes <- c('CD3E', 'TRBC1', 'TRAC', 'TRDC', 'TRGV9', 'NCR1', 'NCR3', 'FCGR3A', 'NCAM1', "GZMB", "GZMA", "GZMK", "GZMH", "PRF1", "GNLY", "KLRD1", "KLRC1", "KLRK1", "ICOS", "PDCD1", "LAG3", "CTLA4", 'TIGIT', 'TCF7') # CD160 and CD244 are not expressed in the XRTCC cluster

# Ensure 'seurat_clusters' is set as the identity class
Idents(sub) <- 'seurat_clusters'

# Get the expression data for the genes of interest
expr_data <- as.matrix(sub@assays$RNA$scale.data[genes, ])

# Extract cluster information for annotation
cluster_info <- as.data.frame(Idents(sub))
colnames(cluster_info) <- "Cluster"

# Extract 'XRTCC' information from metadata
xrtcc_info <- sub@meta.data$XRTCC_cdr3
xrtcc_info[xrtcc_info == ''] <- 'not XDRTCC' 
xrtcc_info <- as.factor(xrtcc_info)
xrtcc_info <- as.data.frame(xrtcc_info) 
colnames(xrtcc_info) <- "XRDTCC CDR3"

# Combine cluster information and 'XRTCC' information
annotation_info <- cbind(cluster_info, xrtcc_info)

# Reorder columns (cells) by cluster
cell_order <- order(annotation_info$Cluster)
expr_data_ordered <- expr_data[, cell_order]
annotation_info_ordered <- annotation_info[cell_order, , drop = FALSE]

# Define the custom color gradient
custom_colors <- colorRampPalette(c("cornflowerblue", "coral3"))(100)

# Generate annotation colors dynamically for clusters
unique_clusters <- unique(annotation_info_ordered$Cluster)
cluster_colors <- setNames(c('#F8766D', '#CD9600', '#7CAE00', '#00C19A', '#00B8E7', '#C77CFF', '#FF61CC'), unique_clusters)

unique_xrtcc <- unique(annotation_info_ordered$`XRDTCC CDR3`)
xrtcc_colors <- setNames(c('white', '#F8766D', '#CD9600', '#7CAE00', '#00C19A', '#00B8E7', '#C77CFF'), unique_xrtcc)

# Combine all annotation colors
anno_colors <- list(Cluster = cluster_colors, 'XRDTCC CDR3' = xrtcc_colors)

# Create the heatmap with the custom color gradient and column annotations
pheatmap(expr_data_ordered,
         annotation_col = annotation_info_ordered,
         cluster_cols = FALSE,
         cluster_rows = FALSE,
         show_rownames = TRUE,
         show_colnames = FALSE,
         main = "Heatmap of Selected Genes by Cluster",
         color = custom_colors,
         annotation_colors = anno_colors)
```

Figure 6C, 6E
```{r}
AB = WhichCells(sub, expression = TRBC1 > 0.5 & CD3E > 0.5 & TRDC < 0.5, slot = 'data')
DimPlot(sub, cells.highlight = AB)

yd = WhichCells(sub, expression = CD3E > 0.5 & (TRDC > 0.5), slot = 'data')
DimPlot(sub, cells.highlight = yd)

NK = WhichCells(sub, expression = CD3E < 0.5 & (NCAM1 > 0.5 | FCGR3A > 0.5), slot = 'data')
DimPlot(sub, cells.highlight = NK)

# Quantify types of cells per cluster
meta$type = ''
meta[AB,]$type = 'AB'
meta[NK,]$type = 'NK'
meta[yd,]$type = 'yd'

sub@meta.data = meta
DimPlot(sub, group.by = 'type')

table(meta$type)

# Fig 6E
# Define the genes of interest
library(ggpubr)
library(cowplot) # For combining plots into a grid

DefaultAssay(sub) = 'RNA'

comparison_list <- list(c('1', '5'), c('6', '5'), c('3', '5'), c('4', '5')) # List of comparisons
genes <- c('CD3E', 'TRAC', 'TRDC', 'TRGV9', "GZMB", "GZMA", "GZMH", "PRF1", "GNLY", "KLRD1", "KLRC1", "KLRK1") # List of gene sets
plots_adjusted <- list() # Initialize a list to store plots

# Generate the violin plot for the current set of genes
for(i in seq_along(genes)){
  plots <- VlnPlot(sub, features = genes[i]) + NoLegend()
  y_limits <- ggplot_build(plots)$layout$panel_scales_y[[1]]$range$range
  
  # Adjust ylim and add non-overlapping p-value labels
  plots_adjusted[[i]] = plots + 
    stat_compare_means(comparisons = comparison_list, 
                                 label = "p.signif", 
                                 method = 'wilcox.test',
                                 step.increase = 0.2, # Prevents the p-value symbols from overlapping on graph
                                 tip.length = 0) + # Length of the downward tip of each line for each comparison on the graph
          ylim(y_limits[1], y_limits[2] + 4) + # Extends y-axis of each graph to accommodate the p-value signs
          theme(axis.title.x = element_blank(), # Hide x-axis title
                axis.title.y = element_blank()) # Hide y-axis title
}
 
# print each plot individually, copy paste into ppt
combined_plot <- plot_grid(plotlist = plots_adjusted, ncol = 6, nrow = 2)
print(combined_plot) # Display the combined plots
```

1. XRTCC
Save XRTCCs in dat object
```{r}
saveRDS(dat, file = 'integrated3.rds')
```

Comparison
```{r}
path = 'Z:/CCTI_SYKES/EXP FF7- XKLT-001 Exp/single cell RNA analysis XKLT001 Analysed data/'
old = readRDS(paste0(path, 'integrateddataRDS3'))
```


```{r}
path = 'Z:/CCTI_USERS/Nathan Suek/2. Collaborations/1. Farshid/2. Decedent xeno/'
old = readRDS(paste0(path, 'c4_tcells_only.rds')) 

genes = c("GZMB", "GZMA", "GZMK", "GZMH", "PRF1", "PDCD1", "LAG3","ICOS", "CTLA4")

# Get the expression data for the genes of interest
expr_data <- as.matrix(sub@assays$SCT$scale.data[genes,])
seurat_obj = sub
```

1. Subcluster xRTCCs
```{r}
# sub
sub = subset(dat, cells = highlight)
DimPlot(sub)
sub@assays$SCT = NULL
sub@assays$integrated = NULL
sub@meta.data = sub@meta.data[,-c(9:108)]
DefaultAssay(sub) = 'RNA'
sub = JoinLayers(sub)
sub = NormalizeData(sub)
sub = ScaleData(sub)

s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
sub <- CellCycleScoring(sub, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
sub <- RunPCA(sub, features = c(s.genes, g2m.genes))
DimPlot(sub)

sub = SCTransform(sub, return.only.var.genes = F, verbose = T, conserve.memory = F)
sub = RunPCA(sub, features = VariableFeatures(object = sub))
sub = RunUMAP(sub, dims = 1:30, verbose = FALSE, umap.method = 'umap-learn', metric = 'correlation')
sub = FindNeighbors(sub, dims = 1:30, verbose = F)
sub = FindClusters(sub, resolution = seq(0.01,2.2, by= 0.2), verbose = F, algorithm = 1)

DoHeatmap(sub, features = jianing_genes1, assay = 'RNA') + scale_fill_gradientn(colors = c("cornflowerblue","coral3"))
DoHeatmap(sub, features = jianing_genes2, assay = 'RNA') + scale_fill_gradientn(colors = c("cornflowerblue","coral3"))

markers = FindAllMarkers(sub)
top = markers %>% group_by(cluster) %>% top_n(6, avg_log2FC)
DoHeatmap(sub, features = top$gene, assay = 'RNA') + scale_fill_gradientn(colors = c("cornflowerblue","coral3"))
```

2. pheatmap
Edit: 7.14.24 I didn't look at this too carefully but I think at the time I was trying to do hierarchical clustering within each seurat cluster cause I was seeing patterns between cell expression 
```{r}
genes = c("GZMB", "GZMA", "GZMK", "GZMH", "PRF1", "PDCD1", "LAG3","ICOS", "CTLA4")

# Get the expression data for the genes of interest
expr_data <- as.matrix(sub@assays$SCT$scale.data[genes,])
seurat_obj = sub

# Get the cluster identities
Idents(seurat_obj) = seurat_obj@meta.data[,'SCT_snn_res.2.01']
clusters <- Idents(seurat_obj)
meta = seurat_obj@meta.data

# Initialize a list to store clustered data for each cluster
clustered_data <- list()
cluster_annotations <- c()  # To store cluster annotations for each cell

# Loop through each cluster, perform hierarchical clustering, and store the results
for (cluster in unique(clusters)) {
  # Subset the data for the current cluster
  cluster_cells <- WhichCells(seurat_obj, idents = cluster)
  cluster_expr_data <- expr_data[, cluster_cells]
  
  # Perform hierarchical clustering
  dist_matrix <- dist(t(cluster_expr_data))
  hclust_result <- hclust(dist_matrix)
  
  # Order the cells based on the clustering result
  ordered_cells <- cluster_cells[hclust_result$order]
  
  # Store the ordered expression data
  clustered_data[[cluster]] <- cluster_expr_data[, ordered_cells]
  cluster_annotations <- c(cluster_annotations, rep(cluster, length(ordered_cells)))
}

# Combine the clustered data from all clusters
combined_data <- do.call(cbind, clustered_data)

# Create an annotation data frame for the clusters
annotation <- data.frame(Cluster = factor(cluster_annotations, levels = unique(cluster_annotations)))
rownames(annotation) <- colnames(combined_data)

annotation$DRTCC = '0'
annotation$DRTCC = meta[match(rownames(annotation), rownames(meta)), "DRTCC"]
annotation$DRTCC <- factor(annotation$DRTCC, levels = unique(annotation$DRTCC))

annotation$cdr3 = meta[match(rownames(annotation), rownames(meta)), 'TCRb_1']
annotation[annotation$DRTCC == 'nonDRTCC','cdr3'] = '0'
annotation$cdr3 <- factor(annotation$cdr3, levels = unique(annotation$cdr3))

# Define annotation colors for clusters
cluster_levels <- levels(annotation$Cluster)
annotation_colors <- list(Cluster = setNames(rainbow(length(cluster_levels)), cluster_levels))
drtcc_levels <- levels(annotation$DRTCC)
annotation_colors2 <- list(DRTCC = setNames(c('lightgray', 'red'), drtcc_levels))
cdr3_levels <- levels(annotation$cdr3)
annotation_colors3 <- list(cdr3 = setNames(c('lightgray', 'yellow', 'green', 'blue', 'purple', 'brown', 'red'), cdr3_levels))

# Define a custom color palette for the heatmap expression values
my_palette <- colorRampPalette(c("cornflowerblue","coral3"))(100)

additional_genes = c('CD3E', 'CD3D', 'CD3G', 'CD4', 'CD8A')
# Add some genes
final <- rbind(combined_data, as.matrix(sub@assays$SCT$scale.data[additional_genes, ]))

# Plot the heatmap with annotations and hierarchical clustering within each cluster
pheatmap(final, cluster_rows = FALSE, cluster_cols = FALSE, show_colnames = FALSE, show_rownames = TRUE,
         annotation_col = annotation, annotation_colors = c(annotation_colors, annotation_colors2, annotation_colors3), color = my_palette)
```

Figure 4
Plot T cells above and XRTCCs below
```{r}
setwd('Z:/CCTI_USERS/Nathan Suek/2. Collaborations/1. Farshid/2. Decedent xeno/2. Paper Analysis/1. Master Analysis')
dat = readRDS('integrated5.rds')

# Change POD61 rejection status to 'non-rejecting'
meta = dat@meta.data %>% mutate(rejection = if_else(POD == 'POD 61' & rejection == 'rejecting', 'non-rejecting', rejection))
meta = dat@meta.data %>% mutate(POD = case_when(
  POD == 'POD 14K' ~ 'POD14K', 
  POD == 'POD 14L' ~ 'POD14L', 
  POD == 'POD 28' ~ 'POD28K', 
  POD == 'POD 33' ~ 'POD33K',
  POD == 'POD 61' ~ 'POD61K'
))
dat@meta.data = meta

# Figure 4B
dat$seurat_clusters = Idents(dat)
DimPlot_scCustom(dat, split.by = 'POD', num_columns = 5, ggplot_default_colors =  T, label = F)
DimPlot_scCustom(dat, label = F, ggplot_default_colors = T) + ggtitle('Total') + theme(plot.title = element_text(hjust = 0.5))

table(dat$XRTCC, dat$seurat_clusters)

dat = ScaleData(dat, features = rownames(dat))

markers = FindAllMarkers(dat)
top = markers %>% group_by(cluster) %>% top_n(20, wt = avg_log2FC)
#DoHeatmap(dat, features = top$gene) + scale_fill_gradientn(colors = c("cornflowerblue","coral3"))

# Extract data for the heatmap
heatmap_data <- as.matrix(GetAssayData(dat, slot = "scale.data"))

# Extract cell cluster information
clusters <- Idents(dat)
col_order <- order(clusters)
heatmap_data <- heatmap_data[, col_order]

# Create annotation for clusters
annotation_col <- data.frame(Cluster = clusters[col_order])
rownames(annotation_col) <- colnames(heatmap_data)

# Select genes and create a vector of row labels with only selected genes labeled
selected_genes <- c('FOXP3', 'SELL', 'NOG', 'FAS', 'CCR4')
row_labels <- ifelse(rownames(heatmap_data) %in% selected_genes, rownames(heatmap_data), "")

# Create the heatmap with pheatmap
#pheatmap(heatmap_data, color = colorRampPalette(c("cornflowerblue", "coral3"))(100), annotation_col = annotation_col, cluster_cols = F, cluster_rows = F, annotation_names_col = TRUE, show_colnames = F)






library(ComplexHeatmap)
library(circlize)

jianing_genes1 = c("CD79A", "CD74", 'MS4A1', 'CD14', 'ITGAM', 'ITGAX',"CD3E", "CD3D", "CD3G", "CD4", "CD8A", 'CD8B', 'CD4', 'NCAM1', 'NCR1', "TRBC1","TRDC", 'TRGV9', 'IGHM', "CD69", "ITGAE", "ITGA1", "CXCR6", "RUNX3", "PRDM1", "CCR7", "KLF2", "S1PR1", "SELL", "MKI67", "CD28","TMIGD2", "TIGIT","TCF7", "RORA", "RORC", "AHR", "IL17RA","CCL20", "TNF", "IFNG", "IFNGR1","TBX21", "EOMES", "GATA3")

jianing_genes2 = c("BCL6", "CXCR5", "PDCD1", "ICOS", "IL2RA", "CXCL13", "IL21", "IL6", "STAT3", "IRF4", "IL21R", "IL6R", "BATF", "STAT5A", "MAF", "GZMB", "GZMA","PRF1","GNLY", "KLRD1", "KLRC1", "KLRK1","FOXP3", "CTLA4","IL10","TGFB1", "IL2","NR4A1", "SATB1", "EGR1", "IL4I1", "NFATC1","JUN", "FOS","NFKBIA", "NFKBID","CD83", "PTPN22", "CXCR4", "BTG2", "HLA-A", "HLA-B", "HLA-C", "HLA-DRB1", "HLA-DPB1", "HLA-DQB1", "LYN", "CCL4")

genes = c('CD2', 'CD3D', 'CD3E', 'CD3G', 'CD4', 'CD8A', 'TRDC', 'CD28', 'TBX21', 'GATA3', 'FOXP3', 'BCL6', 'ASCL2', 'ICOS', 'GNLY', 'CCL5', 'CCL4', 'CXCL8', 'CXCL11', 'GZMB', 'PRF1', 'IL2RA', 'TNF', 'CCR7', 'CD69', 'CD274', 'FCGR1A', 'FCGR1B', 'FCGR2A', 'FCGR2B', 'FCGR3A', 'NCR1', 'NCR3', 'KLRC1', 'KLRC2', 'NCAM1', 'KLRK1', 'CD14', 'CD80', 'CD80', 'CD86', 'ITGAX', 'ITGAM', 'CD163', 'IFNG', 'HLA-A', 'HLA-B', 'HLA-DRB1')

# Assuming heatmap_data, annotation_col, and selected_genes are already defined
# Create the color palette
selected_genes = genes
heatmap_data <- as.matrix(GetAssayData(dat, slot = "scale.data")[selected_genes, ])

clusters <- Idents(dat)
col_order <- order(clusters)
heatmap_data <- heatmap_data[, col_order] # order heatmap by clusters

# Assuming heatmap_data and col_order are defined
clusters <- Idents(dat)[col_order]
rejection_status <- dat$rejection[col_order]

# Combine cluster and rejection status into a data frame
metadata <- data.frame(Cluster = clusters, Rejection = rejection_status)

# Order by Rejection, then by Cluster
ordered_metadata <- metadata %>%
  arrange(Rejection, Cluster)

# Reorder heatmap_data according to the new order
heatmap_data <- heatmap_data[, rownames(ordered_metadata)]
clusters <- ordered_metadata$Cluster
rejection_status <- ordered_metadata$Rejection

colors <- colorRampPalette(brewer.pal(8, "Set3"))(13)
cluster_colors <- setNames(colors, levels(as.factor(clusters)))
colors <- c("yellow", "gold")
rejection_colors = setNames(colors, levels(as.factor(rejection_status)))
ha <- HeatmapAnnotation(Cluster = as.factor(clusters), 
                        Rejection = anno_simple(as.character(rejection_status)), 
                        col = list(Cluster = cluster_colors, Rejection = rejection_colors),
                        annotation_name_side = 'left')

Heatmap(heatmap_data, col = colorRampPalette(c("cornflowerblue", "coral3"))(100), cluster_columns = FALSE, cluster_rows = FALSE, 
              show_column_names = FALSE, top_annotation = ha, show_row_names = T) # Heatmap with all genes labeled



# MAKE HEATMAP WITH SPECIFIC GENES HIGHLIGHTED
# Define selected_rows for annotation
selected_rows <- which(rownames(heatmap_data) %in% c('NCR1', 'CD4', 'CD8A', 'CD8B', 'TRDC')) # Replace some_gene_list with your list of genes for annotation

# Generate the heatmap
Heatmap(heatmap_data, cluster_columns = FALSE, cluster_rows = FALSE, 
              show_column_names = FALSE, top_annotation = ha, show_row_names = FALSE) +
      rowAnnotation(link = anno_mark(at = selected_rows, labels = rownames(heatmap_data)[selected_rows])) # Heatmap with specific genes labeled

"#8DD3C7" "#CFECBB" "#F4F3B9"  "#D2A6B7" "#F5847A"  "#8AB1C9" "#D3B387" "#EABE63" "#BFD767" "#D1D69C" "#FCCDE5"
```

Figure 5
XRTCC figure and heatmap
```{r}
DimPlot_scCustom(dat, split.by = 'POD', group.by = 'XRTCC', num_columns = 5, colors_use = c('gray', 'firebrick', 'blue'), pt.size = 1, order = c(' ', 'XRTCC')) # not sure why it's asking for 3rd color but I think the plot it produces is correct

markers = FindAllMarkers(dat)
top = markers %>% group_by(cluster) %>% top_n(30, wt = avg_log2FC)
top = top %>% filter(cluster == '3' | cluster == '6')

sub = subset(dat, idents = c('3', '6'))
DoHeatmap(sub, features = top$gene, group.colors = c("#CECBD0", "#BD98A2")) + scale_fill_gradientn(colors = c("cornflowerblue","coral3"))

# This version used for Figures
DEenrichRPlot_NS(dat, ident.1 = '3', ident.2 = '6', max.genes = 65, balanced = F, enrich.database = "KEGG_2021_Human")
DEenrichRPlot_NS(dat, ident.1 = '6', ident.2 = '3', max.genes = 65, balanced = F, enrich.database = "KEGG_2021_Human")

# DEenrichPlot: sorts top genes by most statistically significant (DEenrichRPlotNS uses top genes by logFC)
DEenrichRPlot(dat, ident.1 = '3', max.genes = 50, enrich.database = "KEGG_2021_Human")
DEenrichRPlot(dat, ident.1 = '6', max.genes = 50, enrich.database = "KEGG_2021_Human")

# Wiki Pathways
DEenrichRPlot(dat, ident.1 = '3', max.genes = 50, enrich.database = "WikiPathway_2021_Human")
DEenrichRPlot(dat, ident.1 = '6', max.genes = 50, enrich.database = "WikiPathway_2021_Human")

# Cytotoxiciy gene score
# https://www.gsea-msigdb.org/gsea/msigdb/human/geneset/GOBP_POSITIVE_REGULATION_OF_T_CELL_MEDIATED_CYTOTOXICITY.html
geneset = c('MICA','CTSC','LILRB1','RIPK3','RAET1E','IL23R','CTSH','RAET1L','AGER','EMP2','FCGR2B','CADM1','GZMM','NCKAP1L','HFE','HLA-A','HLA-B','HLA-C','HLA-DRA','HLA-DRB1','HLA-E','HLA-F','HLA-G','HLA-H','MR1','HPRT1','HSPA8','RAET1G','IL7R','IL12A','IL12B','IL12RB1','KLRC1','KLRD1','MICB','P2RX7','IL23A','CYRIB','PPP3CB',"PRF1",'CRTAM','B2M','PTPRC','PVR','NECTIN2','RAB27A','CEACAM1','XCL1','TAP2','ULBP3','ULBP2','ULBP1','STX7','FADD','CD1A','CD1B','CD1C','CD1D','CD1E','EBAG9','SLC22A13')
sub = AddModuleScore(sub, features = geneset)
FeaturePlot(sub, features = 'Cluster1', cols = c('white', 'blue'))
```

Figure 6
VDJ Gene Usage of YD T cells 
```{r}
dat = readRDS('integrated4.rds')
yd_genes = c('TRGV10', 'TRGC1', 'TRGV9', 'TRGV7', 'TRGV5', 'TRGV3', 'TRDC') # when I manually scrolled through the genes, this is all the yd related genes present

# Strategy 1: define yd T cells as above (using GEX definition)
Idents(dat) = "POD"
yd_POD14L = WhichCells(dat, expression = CD3E > 0.5 & (TRDC > 0.5), slot = 'data', ident = 'POD 14L')
yd_POD33 = WhichCells(dat, expression = CD3E > 0.5 & (TRDC > 0.5), slot = 'data', ident = 'POD 33')

DimPlot(dat, split.by = 'POD', group.by = 'seurat_clusters', cells = c(yd_POD14L, yd_POD33))

yd_POD14L = subset(dat, cells = yd_POD14L)
yd_POD33 = subset(dat, cells = yd_POD33)

yd_POD14L_sub = yd_POD14L@assays$RNA$data %>% as.data.frame() %>% filter(row.names(yd_POD14L) %in% yd_genes)
yd_POD33_sub = yd_POD33@assays$RNA$data %>% as.data.frame() %>% filter(row.names(yd_POD33) %in% yd_genes)

yd_POD14L_sub = code_binary(yd_POD14L_sub) # this codes any cell expressing any amount of a gene as 1 (and non expressing as 0)
yd_POD33_sub = code_binary(yd_POD33_sub)

yd_POD14L_sub %>% slice(-c(2:3)) %>% rowSums()# %>% prop.table() # TRGV10: 1
yd_POD33_sub %>% slice(-c(2:3)) %>% rowSums()# %>% prop.table() # TRGV10: 0 TRGV9: 0.484 TRGV7: 0.303 TRGV5: 0.1212 TRGV3: 0.0909



# Strategy 2: define yd T cells as any cells that express any yd_genes
yd_POD14L = subset(dat, idents = 'POD 14L')
yd_POD14L = yd_POD14L@assays$RNA$data %>% as.data.frame() %>% filter(row.names(yd_POD14L) %in% yd_genes) %>% select(where(~ sum(.) != 0)) %>% colnames()
yd_POD14L_sub = subset(dat, cells = yd_POD14L)

yd_POD33 = subset(dat, idents = 'POD 33')
yd_POD33 = yd_POD33@assays$RNA$data %>% as.data.frame() %>% filter(row.names(yd_POD33) %in% yd_genes) %>% select(where(~ sum(.) != 0)) %>% colnames()
yd_POD33_sub = subset(dat, cells = yd_POD33)

yd_POD14L_sub = yd_POD14L_sub@assays$RNA$data %>% as.data.frame() %>% filter(row.names(yd_POD14L_sub) %in% yd_genes) %>% code_binary()
yd_POD33_sub = yd_POD33_sub@assays$RNA$data %>% as.data.frame() %>% filter(row.names(yd_POD33_sub) %in% yd_genes) %>% code_binary()

yd_POD14L_sub %>% slice(-c(2:3)) %>% rowSums() %>% prop.table() # TRGV10: 1
yd_POD33_sub %>% slice(-c(2:3)) %>% rowSums() %>% prop.table() # TRGV10: 0 TRGV9: 0.3415 TRGV7: 0.2475 TRGV5: 0.21287 TRGV3: 0.198 
```

GEX of YD at POD14L vs POD33
Figure s10
```{r}
# Using Strategy #1 from above (i.e. GEX definition)
Idents(dat) = "POD"
yd_POD14L = WhichCells(dat, expression = CD3E > 0.5 & (TRDC > 0.5), slot = 'data', ident = 'POD 14L')
yd_POD33 = WhichCells(dat, expression = CD3E > 0.5 & (TRDC > 0.5), slot = 'data', ident = 'POD 33')

dat@meta.data$yd = 0
dat@meta.data[yd_POD14L,]$yd = 'POD14L yd'
dat@meta.data[yd_POD33,]$yd = 'POD33 yd'
sub = subset(dat, yd != 0)



# Figure S9
Idents(sub) = 'yd'
VlnPlot(sub, features = genes, ncol = 5)

# Fig S7
comparison_list <- list(c('POD14L yd', 'POD33 yd')) # List of comparisons
genes = c('CD3E', 'GZMB', 'GZMA', 'GZMH', 'PRF1', 'GNLY', 'KLRD1', 'KLRC1', 'KLRK1')
plots_adjusted <- list() # Initialize a list to store plots


# Generate the violin plot for the current set of genes
for(i in seq_along(genes)){
  plots <- VlnPlot(sub, features = genes[i]) + NoLegend()
  y_limits <- ggplot_build(plots)$layout$panel_scales_y[[1]]$range$range
  
  # Adjust ylim and add non-overlapping p-value labels
  plots_adjusted[[i]] = plots + 
    stat_compare_means(comparisons = comparison_list, 
                                 label = "p.signif", 
                                 method = 'wilcox.test',
                                 step.increase = 0.2, # Prevents the p-value symbols from overlapping on graph
                                 tip.length = 0) + # Length of the downward tip of each line for each comparison on the graph
          ylim(y_limits[1], y_limits[2] + 0.5) + # Extends y-axis of each graph to accommodate the p-value signs
          theme(axis.title.x = element_blank(), # Hide x-axis title
                axis.title.y = element_blank()) # Hide y-axis title
}
 
# print each plot individually, copy paste into ppt
combined_plot <- plot_grid(plotlist = plots_adjusted, ncol = 5, nrow = 2)
combined_plot

# Figure S10A
seurat_obj <- readRDS('integrated4.rds')
highlight = WhichCells(seurat_obj, expression = CD3E > 0.5 & (TRDC > 0.5), slot = 'data')
obj.list <- SplitObject(seurat_obj, split.by = "POD")

lapply(obj.list, function(i){
  DimPlot(i, split.by = 'POD', cells.highlight = highlight, pt.size = 2) + NoLegend() + xlim(-10, 20) + ylim(-10, 25)
})


# Figure S10
Idents(seurat_obj)
sub = subset(seurat_obj, idents = c('3', '6'))

comparison_list <- list(c('3', '6')) # List of comparisons
genes = c('GZMB', 'GZMA', 'GZMK', 'PRF1', 'EOMES', 'PDCD1', 'ICOS', 'TNF', 'IFNG', 'GNLY', 'KLRD1', 'KLRC1')
plots_adjusted <- list() # Initialize a list to store plots

for(i in seq_along(genes)){
  plots <- VlnPlot(sub, features = genes[i]) + NoLegend()
  y_limits <- ggplot_build(plots)$layout$panel_scales_y[[1]]$range$range
  
  # Adjust ylim and add non-overlapping p-value labels
  plots_adjusted[[i]] = plots + 
    stat_compare_means(comparisons = comparison_list, 
                                 label = "p.signif", 
                                 method = 'wilcox.test',
                                 step.increase = 0.2, # Prevents the p-value symbols from overlapping on graph
                                 tip.length = 0) + # Length of the downward tip of each line for each comparison on the graph
          ylim(y_limits[1], y_limits[2] + 0.5) + # Extends y-axis of each graph to accommodate the p-value signs
          theme(axis.title.x = element_blank(), # Hide x-axis title
                axis.title.y = element_blank()) # Hide y-axis title
}
combined_plot <- plot_grid(plotlist = plots_adjusted, ncol = 4, nrow = 3)
combined_plot
```

Figure s5
T cells: CD3E, CD4, CD8A, CD28
NK cells FCGR3A, NCR1, NCAM1 
APCs: CD80, CD86, HLA-DRB1
```{r}
dat = readRDS('integrated5.rds')
dat@meta.data$seurat_clusters = Idents(dat)
#dat = seurat_obj

meta = dat@meta.data
meta$seurat_clusters = Idents(dat)
meta$celltype = ''

# The following definitions are mathematically exclusive
Tcell = WhichCells(dat, expression = CD3E > 0.5 & TRBC1 > 0.5 & TRDC < 0.5)
NK = WhichCells(dat, expression = CD3E < 0.5 & NCR1 > 0.5 & (FCGR3A > 0.5 | NCAM1 > 0.5))
APC = WhichCells(dat, expression = CD3E < 0.5 & NCR1 < 0.5 & (CD14 > 0.5 | ITGAM > 0.5 | ITGAX > 0.5) & (CD80 > 0.5 | CD86 > 0.5 | 'HLA-DRB1' > 0.5))

meta[Tcell,]$celltype = 'T cell'
meta[NK,]$celltype = 'NK'
meta[APC,]$celltype = 'APC'

dat@meta.data = meta

DimPlot(dat, group.by = 'celltype', cols = c('lightgray', '#F8766D', '#7CAE00', '#00BFC4'), pt.size = 1)
write.csv(file = 'temp.csv', prop.table(table(meta$celltype, meta$POD), margin = 2))
```

SingleR
```{r}
seurat_obj = readRDS('integrated4.rds')
# SingleR annotate seurat_obj
library(celldex)
hpca.se <- HumanPrimaryCellAtlasData()
library(SingleR)

hpca.se <- HumanPrimaryCellAtlasData()
sce <- as.SingleCellExperiment(seurat_obj)
pred.main = SingleR(test = sce, assay.type.test = 1, ref = hpca.se, labels = hpca.se$label.main)
pred.fine = SingleR(test = sce, assay.type.test = 1, ref = hpca.se, labels = hpca.se$label.fine)
seurat_obj@meta.data$hpca.main = pred.main$pruned.labels
seurat_obj@meta.data$hpca.fine = pred.fine$pruned.labels

all.markers <- metadata(pred.main)$de.genes
bcell_markers = all.markers$B_cell %>% unlist() %>% unique()
dc_markers = all.markers %>% unlist() %>% unname() %>% unique()

DimPlot(seurat_obj, group.by = 'hpca.main', label = T, repel = T, label.size = 5) + NoLegend() 
DimPlot(seurat_obj, group.by = 'hpca.fine', label = T, repel = T) + NoLegend()


# Cell type annotation: distinguish B cells from DCs in dataset
FeaturePlot(seurat_obj, features = c('CD27', 'CD21', 'IGHM', 'IGHD', 'IGHG1', 'IGHG2', 'IGHG3', 'IGHG4', 'CD93', 'CXCR4', 'BCL6', 'CD38', 'CD1D', 'CXCR5', 'PRDM1', 'CD74'), ncol= 5, order = T) # B cells
FeaturePlot(seurat_obj, features = c('CD14', 'CD16', 'ITGAM', 'ITGAB', 'HLA-DRA', 'HLA-DRB1', 'CD80', 'CD83', 'CD86', 'CD40'), ncol= 4, order = T) # macrophages/DCs
```

Cellchat Seurat Clusters
```{r}
# Load necessary libraries
library(CellChat)

seurat_obj <- readRDS('integrated4.rds')
seurat_obj$seurat_clusters = Idents(seurat_obj)
seurat_obj@meta.data$seurat_clusters <- as.character(seurat_obj@meta.data$seurat_clusters)
seurat_obj@meta.data$seurat_clusters[seurat_obj@meta.data$seurat_clusters == '0'] <- '13' # Substitute '0' with '13'

data.input = seurat_obj@assays$RNA$data # normalized data matrix
meta = seurat_obj@meta.data # a dataframe with rownames containing cell mata data

cellchat <- createCellChat(object = data.input, meta = meta, group.by = "seurat_clusters")

cellchat <- addMeta(cellchat, meta = meta)
cellchat <- setIdent(cellchat, ident.use = "seurat_clusters") # set "labels" as default cell identity
levels(cellchat@idents) # show factor levels of the cell labels
groupSize <- as.numeric(table(cellchat@idents)) # number of cells in each cell group



CellChatDB <- CellChatDB.human # use CellChatDB.mouse if running on mouse data
showDatabaseCategory(CellChatDB)
# Show the structure of the database
dplyr::glimpse(CellChatDB$interaction)

# use all CellChatDB except for "Non-protein Signaling" for cell-cell communication analysis
CellChatDB.use <- subsetDB(CellChatDB)

# set the used database in the object
cellchat@DB <- CellChatDB.use


# subset the expression data of signaling genes for saving computation cost
cellchat <- subsetData(cellchat) # This step is necessary even if using the whole database
future::plan("multisession", workers = 4) # do parallel
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)


cellchat <- computeCommunProb(cellchat, type = "triMean")
cellchat <- filterCommunication(cellchat, min.cells = 10)
cellchat <- computeCommunProbPathway(cellchat)
cellchat <- aggregateNet(cellchat)

groupSize <- as.numeric(table(cellchat@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
groupSize <- as.numeric(table(cellchat@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")

mat <- cellchat@net$weight
par(mfrow = c(3,4), xpd=TRUE)
for (i in 1:nrow(mat)) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}










# (1) show all the significant interactions (L-R pairs) from some cell groups (defined by 'sources.use') to other cell groups (defined by 'targets.use')
netVisual_bubble(cellchat, sources.use = 4, targets.use = c(5:11), remove.isolate = FALSE)
# (2) show all the significant interactions (L-R pairs) associated with certain signaling pathways
netVisual_bubble(cellchat, sources.use = 4, targets.use = c(5:11), signaling = c("CCL","CXCL"), remove.isolate = FALSE)
# (3) show all the significant interactions (L-R pairs) based on user's input (defined by `pairLR.use`)
pairLR.use <- extractEnrichedLR(cellchat, signaling = c("CCL","CXCL","FGF"))
netVisual_bubble(cellchat, sources.use = c(3,4), targets.use = c(5:8), pairLR.use = pairLR.use, remove.isolate = TRUE)




# Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
gg1 <- netAnalysis_signalingRole_scatter(cellchat)
# Signaling role analysis on the cell-cell communication networks of interest
gg2 <- netAnalysis_signalingRole_scatter(cellchat, signaling = c("CXCL", "CCL"))
gg1 + gg2
```

Cellchat Seurat Clusters
```{r}
# Load necessary libraries
library(CellChat)

seurat_obj = seurat_obj[,!is.na(meta$hpca.main)]
meta = seurat_obj@meta.data # a dataframe with rownames containing cell mata data
data.input = seurat_obj@assays$RNA$data # normalized data matrix
data.input = data.input[,!is.na(meta$hpca.main)]

cellchat <- createCellChat(object = data.input, meta = meta, group.by = "hpca.main")

cellchat <- addMeta(cellchat, meta = meta)
cellchat <- setIdent(cellchat, ident.use = "hpca.main") # set "labels" as default cell identity
levels(cellchat@idents) # show factor levels of the cell labels
groupSize <- as.numeric(table(cellchat@idents)) # number of cells in each cell group



CellChatDB <- CellChatDB.human # use CellChatDB.mouse if running on mouse data
showDatabaseCategory(CellChatDB)
# Show the structure of the database
dplyr::glimpse(CellChatDB$interaction)

# use all CellChatDB except for "Non-protein Signaling" for cell-cell communication analysis
CellChatDB.use <- subsetDB(CellChatDB)

# set the used database in the object
cellchat@DB <- CellChatDB.use


# subset the expression data of signaling genes for saving computation cost
cellchat <- subsetData(cellchat) # This step is necessary even if using the whole database
future::plan("multisession", workers = 4) # do parallel
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)


cellchat <- computeCommunProb(cellchat, type = "triMean")
cellchat <- filterCommunication(cellchat, min.cells = 10)
cellchat <- computeCommunProbPathway(cellchat)
cellchat <- aggregateNet(cellchat)

groupSize <- as.numeric(table(cellchat@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")

mat <- cellchat@net$weight
par(mfrow = c(2,2), xpd=TRUE)
for (i in 1:4) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
for (i in 5:8) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
for (i in 9:12) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
for (i in 13:16) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
for (i in 17:20) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
for (i in 21:24) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}

pathways.show <- c("WNT") 
# Hierarchy plot
# Do not define vertex.receiver (assumes all cells can be senders and receivers)
netVisual_aggregate(cellchat, signaling = pathways.show)
# Circle plot
par(mfrow=c(1,1))
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle") 
title(pathways.show)
```

# SYKES REVISIONS 10.15.24
Cellchat: Subset cells (XRTCC, nonXRTCC, APC, NK, yd, B cell) 
Figure 5B
```{r}
setwd('Z:/CCTI_USERS/Nathan Suek/2. Collaborations/1. Farshid/2. Decedent xeno/2. Paper Analysis/1. Master Analysis')
seurat_obj = readRDS('integrated4.rds')
seurat_obj@meta.data$seurat_clusters = Idents(seurat_obj)
seurat_obj@meta.data$subset_labels = ' '

#setwd("/Volumes/CCTI_LABS/CCTI_SYKES/EXP FF7- XKLT-001 Exp/Bulk TCRb sequincing Adaptive Data XKLT001/All MLRs, Official Analysis")
setwd('Z:/CCTI_SYKES/EXP FF7- XKLT-001 Exp/Bulk TCRb sequincing Adaptive Data XKLT001/All MLRs, Official Analysis')
xrtcc_CD8 = read.csv('MasterCD8DRTCCs_AgainstPreTx.csv')
xrtcc_CD4 = read.csv('MasterCD4DRTCCs_AgainstPreTx.csv')
nonxrtcc = read.csv('PreTxNonDRTCCsConservative.csv')

# label all the cells, subset them, run cellchat
meta = seurat_obj@meta.data
tcell = WhichCells(seurat_obj, expression = CD3E > 0.5 & TRBC1 > 0.5 & TRDC < 0.5, slot = 'data')
macrophage_dc = WhichCells(seurat_obj, idents = c(1:7, 9), expression = CD3E < 0.5 & NCR1 < 0.5 & (CD14 > 0.5 | ITGAM > 0.5 | ITGAX > 0.5) & (CD80 > 0.5 | CD86 > 0.5 | 'HLA-DRB1' > 0.5), slot = 'data') # exclude cluster 8 (B cells)
bcell = WhichCells(seurat_obj, idents = '8', slot = 'data')
NK = WhichCells(seurat_obj, expression = CD3E < 0.5 & NCR1 > 0.5 & (FCGR3A > 0.5 | NCAM1 > 0.5), slot = 'data')
yd = WhichCells(seurat_obj, expression = CD3E > 0.5 & TRDC > 0.5, slot = 'data')
DimPlot(seurat_obj, cells.highlight = tcell)
DimPlot(seurat_obj, cells.highlight = macrophage_dc)
DimPlot(seurat_obj, cells.highlight = bcell)
DimPlot(seurat_obj, cells.highlight = NK)
DimPlot(seurat_obj, cells.highlight = yd)

meta$subset_labels = ' '
meta[row.names(meta) %in% tcell,]$subset_labels = 'T cell'
meta[row.names(meta) %in% macrophage_dc,]$subset_labels = 'macrophage_dc'
meta[row.names(meta) %in% bcell,]$subset_labels = 'B cell'
meta[row.names(meta) %in% NK,]$subset_labels = 'NK'
meta[row.names(meta) %in% yd,]$subset_labels = 'yd'
meta[meta$TCRb_1 %in% xrtcc_CD8$Amino.Acid,]$subset_labels = 'XDRTCC'
meta[meta$TCRb_1 %in% xrtcc_CD4$Amino.Acid,]$subset_labels = 'XDRTCC'
meta[meta$TCRb_1 %in% nonxrtcc$Amino.Acid,]$subset_labels = 'non-XDRTCC'
seurat_obj@meta.data = meta
DimPlot(seurat_obj, group.by = 'subset_labels', split.by = 'POD', cols = c('lightgray', scales::hue_pal()(7)), order = c('XDRTCC',  'non-XDRTCC', 'yd', 'NK', 'macrophage_dc', 'B cell', 'T cell', ' '), label = T, repel = T)
DimPlot_scCustom(seurat_obj, group.by = 'subset_labels', split.by = 'POD', colors_use = c('lightgray', scales::hue_pal()(7)), order = c('XDRTCC',  'non-XDRTCC', 'yd', 'NK', 'macrophage_dc', 'B cell', 'T cell', ' '), label = T, repel = T)

"#F8766D" "#E18A00" "#BE9C00" "#8CAB00" "#24B700" "#00BE70" "#00C1AB" "#00BBDA" "#00ACFC" "#8B93FF" "#D575FE" "#F962DD" "#FF65AC"

# Cell chat
Idents(seurat_obj) = 'subset_labels'
sub = subset(seurat_obj, idents = c(" "), invert = TRUE)
#sub = subset(seurat_obj, idents = c("macrophage_dc", "XDRTCC", 'non-XDRTCC'))
data.input <- sub[['RNA']]$data
labels <- sub$subset_labels
meta <- data.frame(labels = labels, row.names = names(labels)) # create a dataframe of the cell labels

cellchat <- createCellChat(object = data.input, meta = meta, group.by = "labels")
CellChatDB <- CellChatDB.human  # Use the human ligand-receptor database (or use CellChatDB.mouse for mouse data)
CellChatDB.use <- subsetDB(CellChatDB)
cellchat@DB <- CellChatDB.use
cellchat <- subsetData(cellchat) # This step is necessary even if using the whole database
future::plan("multisession", workers = 4) # do parallel
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)
cellchat <- computeCommunProb(cellchat, type = "triMean")
cellchat <- filterCommunication(cellchat, min.cells = 10)
cellchat <- computeCommunProbPathway(cellchat)
cellchat <- aggregateNet(cellchat)
cellchat@netP$pathways

groupSize <- as.numeric(table(cellchat@idents))
netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")

pathways.show <- c("NECTIN") 
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle")
netVisual_heatmap(cellchat, signaling = pathways.show, color.heatmap = "Reds")
netAnalysis_contribution(cellchat, signaling = pathways.show)
netVisual_bubble(cellchat, signaling = pathways.show, remove.isolate = FALSE) + theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust = 1), text = element_text(size = 20))
```

Figure 5C
XDRTCC specific gene expression
Figure S7C
```{r}
# Run Figure 5B to get subset_labels
Idents(seurat_obj) = 'subset_labels'
sub = subset(seurat_obj, idents = c('XDRTCC', 'non-XDRTCC'))
meta = sub@meta.data
meta$xdrtcc_clusters = ''
meta = meta %>% mutate(xdrtcc_clusters = ifelse(subset_labels == 'XDRTCC' & seurat_clusters == '0', 'XDRTCC c0',
                                         ifelse(subset_labels == 'XDRTCC' & seurat_clusters == '1', 'XDRTCC c1',
                                         ifelse(subset_labels == 'XDRTCC' & seurat_clusters == '3', 'XDRTCC c3',
                                         ifelse(subset_labels == 'XDRTCC' & seurat_clusters == '4', 'XDRTCC c4',
                                         ifelse(subset_labels == 'XDRTCC' & seurat_clusters == '6', 'XDRTCC c6',
                                         ifelse(subset_labels == 'XDRTCC' & seurat_clusters == '10', 'XDRTCC c10',
                                         ifelse(subset_labels == 'non-XDRTCC' & seurat_clusters == '0', 'non-XDRTCC c0',
                                         ifelse(subset_labels == 'non-XDRTCC' & seurat_clusters == '1', 'non-XDRTCC c1',
                                         ifelse(subset_labels == 'non-XDRTCC' & seurat_clusters == '3', 'non-XDRTCC c3',
                                         ifelse(subset_labels == 'non-XDRTCC' & seurat_clusters == '4', 'non-XDRTCC c4',
                                         ifelse(subset_labels == 'non-XDRTCC' & seurat_clusters == '6', 'non-XDRTCC c6',
                                         ifelse(subset_labels == 'non-XDRTCC' & seurat_clusters == '10', 'non-XDRTCC c10', '')))))))))))))

# Figure S7C
sub@meta.data = meta
Idents(sub) = 'xdrtcc_clusters'
my_levels <- c('non-XDRTCC c0', 'non-XDRTCC c1', 'non-XDRTCC c3', 'non-XDRTCC c4', 'non-XDRTCC c6', 'non-XDRTCC c10', 'XDRTCC c0', 'XDRTCC c1', 'XDRTCC c3', 'XDRTCC c4', 'XDRTCC c6', 'XDRTCC c10')
Idents(sub) = factor(x = Idents(sub), levels = my_levels)
VlnPlot(sub, features = c('GZMA', 'GZMB', 'TNF', 'IFNG', 'PRF1', 'PDCD1', 'LAG3', 'TIGIT', 'EOMES', 'TBX21', 'TGFB1', 'TCF7')) 
```

Figure 4C
Figure S7
```{r}
meta = seurat_obj@meta.data
meta = meta %>% mutate(POD = ifelse(POD == 'POD 14K', 'POD14K', 
                                    ifelse(POD == 'POD 14L', 'POD14L', 
                                           ifelse(POD == "POD 28", "POD28K",
                                                  ifelse(POD == "POD 33", "POD33K",
                                                         ifelse(POD == 'POD 61', 'POD61K', POD))))))
seurat_obj@meta.data = meta
DimPlot_scCustom(seurat_obj, split.by = 'POD', num_columns = 5, colors_use = scales::hue_pal()(13)) 

highlight = seurat_obj@meta.data %>% filter(XRTCC_cdr3 == 'CASSVTGTSGRYEQFF') %>% row.names()
highlight = seurat_obj@meta.data %>% filter(XRTCC_cdr3 == 'CASSISSYNEQFF') %>% row.names()
obj.list <- SplitObject(seurat_obj, split.by = "POD")
DimPlot(obj.list[[1]], split.by = 'POD', cells.highlight = highlight, pt.size = 2) + NoLegend() + xlim(-10, 20) + ylim(-10, 25)
```

Figure 5A
Figure S6A, B: labeling XDRTCCs as indirect, direct, or undefined
```{r}
setwd('Z:/CCTI_USERS/Nathan Suek/2. Collaborations/1. Farshid/2. Decedent xeno/2. Paper Analysis/1. Master Analysis')
seurat_obj = readRDS('integrated4.rds')
meta = seurat_obj@meta.data

setwd('Z:/CCTI_SYKES/EXP FF7- XKLT-001 Exp/Bulk TCRb sequincing Adaptive Data XKLT001/All MLRs, Official Analysis')
xrtcc_CD8 = read.csv('MasterCD8DRTCCs_AgainstPreTx.csv')
xrtcc_CD4 = read.csv('MasterCD4DRTCCs_AgainstPreTx.csv')
nonxrtcc = read.csv('PreTxNonDRTCCsConservative.csv')

meta = meta %>% rename(XDRTCC = XRTCC)
meta[meta$TCRb_1 %in% xrtcc_CD8$Amino.Acid,]$XDRTCC = 'XDRTCC'
meta[meta$TCRb_1 %in% xrtcc_CD4$Amino.Acid,]$XDRTCC = 'XDRTCC'
meta[meta$TCRb_1 %in% nonxrtcc$Amino.Acid,]$XDRTCC = 'non-XDRTCC'
seurat_obj@meta.data = meta
DimPlot_scCustom(seurat_obj, split.by = 'POD', group.by = 'XDRTCC', num_columns = 5, colors_use = c('lightgray', 'blue', 'red'), pt.size = 1, order = T) 
Idents(seurat_obj) = 'XDRTCC'
DimPlot_scCustom(seurat_obj, split.by = 'XDRTCC', num_columns = 3, colors_use = c('lightgray', 'blue', 'red'), pt.size = 1, order = T, label = F) 

# Generate list of Indirect and Direct CD8 CD4
setwd('Z:/CCTI_SYKES/EXP FF7- XKLT-001 Exp/Bulk TCRb sequincing Adaptive Data XKLT001/All MLRs, Official Analysis')
xdrtcc_CD8 = read.csv('MasterCD8DRTCCs_AgainstPreTx.csv')
xdrtcc_CD4 = read.csv('MasterCD4DRTCCs_AgainstPreTx.csv')
nondxrtcc = read.csv('PreTxNonDRTCCsConservative.csv')

xdrtcc_CD4 = xdrtcc_CD4 %>% mutate(response = ifelse(
  (XKLT001.POD49.MLR.Rcipient.PBMC.vs.dodnor.Pig.PBMC.CD4CFSElow_TCRB > 0 | XKLT001_Direct_LN_v_pig_DC_CD4_CFSE_low_TCRB > 0 |
     XKLT001_POD_28_Direct_MLR_CD4_CFSE_low_TCRB > 0 | XKLT001_Direct_PBMC_v_pig_DC_CD4_CFSE_low_TCRB > 0) &
    XKLT001_Indirect_MLR_CD4_CFSE_low_TCRB == 0, 'CD4 Direct', ifelse(
      (XKLT001.POD49.MLR.Rcipient.PBMC.vs.dodnor.Pig.PBMC.CD4CFSElow_TCRB == 0 | XKLT001_Direct_LN_v_pig_DC_CD4_CFSE_low_TCRB == 0 |
         XKLT001_POD_28_Direct_MLR_CD4_CFSE_low_TCRB == 0 & XKLT001_Direct_PBMC_v_pig_DC_CD4_CFSE_low_TCRB == 0) &
        XKLT001_Indirect_MLR_CD4_CFSE_low_TCRB > 0, 'CD4 Indirect', 'CD4 Undefined')))

xdrtcc_CD8 = xdrtcc_CD8 %>% mutate(response = ifelse(
  (XKLT001.POD49.MLR.Rcipient.PBMC.vs.dodnor.Pig.PBMC.CD8CFSElow_TCRB > 0 | XKLT001_Direct_LN_v_pig_DC_CD8_CFSE_low_TCRB > 0 |
     XKLT001_POD_28_Direct_MLR_CD8_CFSE_low_TCRB > 0 | XKLT001_Direct_PBMC_v_pig_DC_CD8_CFSE_low_TCRB > 0) &
    XKLT001_Indirect_MLR_CD8_CFSE_low_TCRB == 0, 'CD8 Direct', ifelse(
      (XKLT001.POD49.MLR.Rcipient.PBMC.vs.dodnor.Pig.PBMC.CD8CFSElow_TCRB == 0 | XKLT001_Direct_LN_v_pig_DC_CD8_CFSE_low_TCRB == 0 |
         XKLT001_POD_28_Direct_MLR_CD8_CFSE_low_TCRB == 0 & XKLT001_Direct_PBMC_v_pig_DC_CD8_CFSE_low_TCRB == 0) &
        XKLT001_Indirect_MLR_CD8_CFSE_low_TCRB > 0, 'CD8 Indirect', 'CD8 Undefined'))) 

# Determine if XDRTCC in kidney are from indirect or direct
meta = seurat_obj@meta.data
meta = meta %>% rownames_to_column('rownmame') %>% left_join(xdrtcc_CD4 %>% select(Amino.Acid, response) %>% rename(response_CD4 = response), by = c("TCRb_1" = "Amino.Acid")) %>% left_join(xdrtcc_CD8 %>% select(Amino.Acid, response) %>% rename(response_CD8 = response), by = c("TCRb_1" = "Amino.Acid")) %>% column_to_rownames('rownmame')
meta = meta %>% mutate(response = case_when(
  is.na(response_CD4) ~ response_CD8, 
  is.na(response_CD8) ~ response_CD4))
meta = meta %>% select(-c(response_CD4, response_CD8))  
meta[is.na(meta$response),]$response = ''
seurat_obj@meta.data = meta

DimPlot(seurat_obj, split.by = 'POD', group.by = 'response', cols = c('lightgray', "#00BE67", "#C77CFF", "#FDE725FF", "#F8766D", "brown", "#00B4F0") , order = c('CD8 Undefined', 'CD4 Indirect', 'CD8 Indirect', 'CD4 Direct', 'CD8 Direct', ''), pt.size = 1.5)
DimPlot_scCustom(seurat_obj, split.by = 'POD', group.by = 'response', colors_use = c('lightgray', "#00BE67", "#C77CFF", "#FDE725FF", "#F8766D", "brown", "#00B4F0") , order = c('CD8 Undefined', 'CD4 Indirect', 'CD8 Indirect', 'CD4 Direct', 'CD8 Direct', ''), pt.size = 1.5, num_columns = 5)

"#440154FF" "#482878FF" "#3E4A89FF" "#31688EFF" "#26828EFF" "#1F9E89FF" "#35B779FF" "#6DCD59FF" "#B4DE2CFF" "#FDE725FF"

"#F8766D" "#E18A00" "#BE9C00" "#8CAB00" "#24B700" "#00BE70" "#00C1AB" "#00BBDA" "#00ACFC" "#8B93FF" "#D575FE" "#F962DD" "#FF65AC"
```

Figure S11
Add VlnPlots for XDRTCC by cluster
```{r}
sub = readRDS('sub_c3.rds')
sub$seurat_clusters = Idents(sub)
DefaultAssay(sub) = 'RNA'

meta = sub@meta.data
meta = meta %>% mutate(xdrtcc_subc3_clusters = case_when(
  XRTCC == 'XRTCC' & seurat_clusters == '0' ~ 'XDRTCC c0',
  XRTCC == 'XRTCC' & seurat_clusters == '2' ~ 'XDRTCC c2',
  XRTCC == 'XRTCC' & seurat_clusters == '5' ~ 'XDRTCC c5',
  TRUE ~ ''))
meta = meta %>% mutate(subc3_clusters = case_when(
  XRTCC == 'XRTCC' & seurat_clusters == '0' ~ 'XDRTCC c0',
  XRTCC == 'XRTCC' & seurat_clusters == '2' ~ 'XDRTCC c2',
  XRTCC == 'XRTCC' & seurat_clusters == '5' ~ 'XDRTCC c5',
  XRTCC == '' & seurat_clusters == '0' ~ 'c0',
  XRTCC == '' & seurat_clusters == '2' ~ 'c2',
  XRTCC == '' & seurat_clusters == '5' ~ 'c5',
  XRTCC == '' & seurat_clusters == '1' ~ 'c1',
  XRTCC == '' & seurat_clusters == '3' ~ 'c3',
  XRTCC == '' & seurat_clusters == '4' ~ 'c4',
  XRTCC == '' & seurat_clusters == '6' ~ 'c6',
  TRUE ~ ''))
sub@meta.data = meta
DimPlot(sub, group.by = 'xdrtcc_subc3_clusters', cols = c('lightgray', scales::hue_pal()(3)), order = c('XDRTCC c0', 'XDRTCC c2', 'XDRTCC c5', ''))

my_levels <- c('c0','c2', 'c5', 'XDRTCC c0', 'XDRTCC c2', 'XDRTCC c5', 'c1', 'c3', 'c4', 'c6')
Idents(sub) = 'subc3_clusters'
Idents(sub) = factor(x = Idents(sub), levels = my_levels)
VlnPlot(sub, features = c('GZMB', 'GZMA', 'GZMH', 'PRF1', 'PDCD1', 'LAG3', 'EOMES', 'CTLA4', 'TIGIT', 'TCF7', 'GNLY', 'KLRD1'), ncol = 6)
```

































Pig and Human Analysis
```{r}
hist_data <- data.frame(Sum = colSums(temp)+1)
ggplot(hist_data, aes(x = Sum)) +
  geom_histogram(bins = 20) +
  scale_y_log10() +
  labs(title = "Histogram with Logarithmic Y-axis", x = "Sum", y = "Frequency") +
  theme_minimal()

pig = hist_data %>% filter(Sum > 500) %>% row.names()
dat = readRDS('integrateddataRDS5')
DimPlot(dat, cells.highlight = pig) # those cells are not present


human_rows = colSums(RNA[1:18462,])
pig_rows = colSums(RNA[18463:24727,])
scatter_data <- data.frame(X = human_rows, Y = pig_rows)
ggplot(scatter_data, aes(x = X, y = Y)) +
  geom_point(color = "blue", alpha = 0.7) +
  geom_abline(intercept = 0, slope = 1, color = "red", linetype = "dashed", size = 1) +
  labs(title = "Scatter Plot", x = "Sum of Human Genes", y = "Sum of Pig Genes") +
  theme_minimal()

scatter_data$Position <- ifelse(scatter_data$Y > scatter_data$X, "Above", "Below")
table(scatter_data$Position)

pig <- rownames(scatter_data[scatter_data$Position == "Above", ])
human <- rownames(scatter_data[scatter_data$Position == "Below", ])

print(pig)
print(human)
```

Hybrid Pig Human Ref
```{r}
path = 'Z:/CCTI_USERS/Nathan Suek/2. Collaborations/1. Farshid/2. Decedent xeno/2. Paper Analysis/1. Master Analysis/0. Archived/24.12.21 Original Master Analysis - including human and pig cells/'
dat = readRDS(paste0(path,'integrated5.rds'))
meta = dat@meta.data

setwd('Z:/CCTI_SYKES/EXP FF7- XKLT-001 Exp/Bulk TCRb sequincing Adaptive Data XKLT001/All MLRs, Official Analysis')
xdrtcc_CD8 = read.csv('MasterCD8DRTCCs_AgainstPreTx.csv')
xdrtcc_CD4 = read.csv('MasterCD4DRTCCs_AgainstPreTx.csv')
nonxrtcc = read.csv('PreTxNonDRTCCsConservative.csv')

meta = dat@meta.data
meta = meta %>% rename(XDRTCC = XRTCC)
meta[meta$TCRb_1 %in% xdrtcc_CD8$Amino.Acid,]$XDRTCC = 'XDRTCC'
meta[meta$TCRb_1 %in% xdrtcc_CD4$Amino.Acid,]$XDRTCC = 'XDRTCC'
meta[meta$TCRb_1 %in% nonxrtcc$Amino.Acid,]$XDRTCC = 'non-XDRTCC'
dat@meta.data = meta
DimPlot(dat, split.by = 'XDRTCC', group.by = 'XDRTCC')
DimPlot(dat, split.by = 'POD', group.by = 'XDRTCC', cols = c('lightgray', 'blue', 'red'))
XDRTCC_count = meta %>% filter(XDRTCC == 'XDRTCC') %>% summarise(count = n())


path = 'Z:/CCTI_USERS/Nathan Suek/2. Collaborations/1. Farshid/2. Decedent xeno/2. Paper Analysis/1. Master Analysis/3. Pig RNAseq Analysis/8. Hybrid Ref v7 - correct GEX original files/'
pig_cells = read.csv(paste0(path,'pig_cells_withPOD61.csv'))
remove = intersect(row.names(meta), pig_cells$pig_cells)

sub = subset(dat, cells = remove, invert = T)
meta2 = sub@meta.data

DimPlot(sub)
sub = RenameIdents(sub, '6' = '5', '7' = '6', '8' = '7', '9' = '8', '10' = '9', '11' = '10', '12' = '11')
my_levels <- c('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11')
Idents(sub) = factor(x = Idents(dat), levels = my_levels)
Idents(dat) = 'POD'
dat = RenameIdents(dat, 'POD 14K' = 'POD14K', 'POD 14L' = 'POD14L', 'POD 28' = 'POD28K', 'POD 33' = 'POD33K', 'POD 61' = 'POD61K')
dat$POD = Idents(dat)
saveRDS(dat, file = 'integrated6_withoutPigCells.rds')

setwd('Z:/CCTI_SYKES/EXP FF7- XKLT-001 Exp/Bulk TCRb sequincing Adaptive Data XKLT001/All MLRs, Official Analysis')
xrtcc_CD8 = read.csv('MasterCD8DRTCCs_AgainstPreTx.csv')
xrtcc_CD4 = read.csv('MasterCD4DRTCCs_AgainstPreTx.csv')
nonxrtcc = read.csv('PreTxNonDRTCCsConservative.csv')

meta = sub@meta.data
meta[meta$TCRb_1 %in% xrtcc_CD8$Amino.Acid,]$XDRTCC = 'XDRTCC'
meta[meta$TCRb_1 %in% xrtcc_CD4$Amino.Acid,]$XDRTCC = 'XDRTCC'
meta[meta$TCRb_1 %in% nonxrtcc$Amino.Acid,]$XDRTCC = 'non-XDRTCC'
sub@meta.data = meta
DimPlot(sub, split.by = 'XDRTCC', group.by = 'XDRTCC')
DimPlot(sub, split.by = 'POD', group.by = 'XDRTCC', cols = c('lightgray', 'blue', 'red'))
XDRTCC_count = meta %>% filter(XDRTCC == 'XDRTCC') %>% summarise(count = n())
```

Figure 4A
```{r}
dat = readRDS('integrated6_withoutPigCells.rds')

ggplotColours <- function(n = 6, h = c(0, 360) + 15){
  if ((diff(h) %% 360) < 1) h[2] <- h[2] - 360/n
  hcl(h = (seq(h[1], h[2], length = n)), c = 100, l = 65)
}
color_list <- ggplotColours(n=12)

DimPlot(dat) 
DimPlot_scCustom(dat, split.by = 'POD', colors_use = color_list, num_columns = 5, label = F) 
prop.table(table(dat$seurat_clusters2, dat$POD), margin = 2)
```

Figure 4B
```{r}
library(ComplexHeatmap)
library(circlize)

genes = c('CD2', 'CD3D', 'CD3E', 'CD3G', 'CD4', 'CD8A', 'TRDC', 'CD28', 'TBX21', 'GATA3', 'FOXP3', 'BCL6', 'ASCL2', 'ICOS', 'GNLY', 'CCL5', 'CCL4', 'CXCL8', 'CXCL11', 'GZMB', 'PRF1', 'IL2RA', 'TNF', 'CCR7', 'CD69', 'CD274', 'FCGR1A', 'FCGR1B', 'FCGR2A', 'FCGR2B', 'FCGR3A', 'NCR1', 'NCR3', 'KLRC1', 'KLRC2', 'NCAM1', 'KLRK1', 'CD14', 'CD80', 'CD80', 'CD86', 'ITGAX', 'ITGAM', 'CD163', 'IFNG', 'HLA-A', 'HLA-B', 'HLA-DRB1')

# Assuming heatmap_data, annotation_col, and selected_genes are already defined
# Create the color palette
selected_genes = genes
dat = ScaleData(dat, features = rownames(dat))
heatmap_data <- as.matrix(GetAssayData(dat, slot = "scale.data")[selected_genes, ])

clusters <- Idents(dat)
col_order <- order(clusters)
heatmap_data <- heatmap_data[, col_order] # order heatmap by clusters

# Assuming heatmap_data and col_order are defined
clusters <- Idents(dat)[col_order]
rejection_status <- dat$rejection[col_order]

# Combine cluster and rejection status into a data frame
metadata <- data.frame(Cluster = clusters, Rejection = rejection_status)

# Order by Rejection, then by Cluster
ordered_metadata <- metadata %>%
  arrange(Rejection, Cluster)

# Reorder heatmap_data according to the new order
heatmap_data <- heatmap_data[, rownames(ordered_metadata)]
clusters <- ordered_metadata$Cluster
rejection_status <- ordered_metadata$Rejection

# Create a palette for clusters
ggplotColours <- function(n = 6, h = c(0, 360) + 15){
  if ((diff(h) %% 360) < 1) h[2] <- h[2] - 360/n
  hcl(h = (seq(h[1], h[2], length = n)), c = 100, l = 65)
}
color_list <- ggplotColours(n=12)

colors <- color_list
cluster_colors <- setNames(colors, levels(as.factor(clusters)))
colors <- c('cornflowerblue', 'tomato')
rejection_colors = setNames(colors, levels(as.factor(rejection_status)))

ha <- HeatmapAnnotation(Cluster = as.factor(clusters), 
                        Rejection = as.factor(rejection_status), 
                        col = list(Cluster = cluster_colors, Rejection = rejection_colors),
                        annotation_name_side = 'left')


Heatmap(heatmap_data, col = colorRampPalette(c("cornflowerblue", "coral3"))(100), cluster_columns = FALSE, cluster_rows = FALSE, 
              show_column_names = FALSE, top_annotation = ha, show_row_names = T) # Heatmap with all genes labeled
```

Figure 5A
```{r}
setwd('Z:/CCTI_SYKES/EXP FF7- XKLT-001 Exp/Bulk TCRb sequincing Adaptive Data XKLT001/All MLRs, Official Analysis')
xrtcc_CD8 = read.csv('MasterCD8DRTCCs_AgainstPreTx.csv')
xrtcc_CD4 = read.csv('MasterCD4DRTCCs_AgainstPreTx.csv')
nonxrtcc = read.csv('PreTxNonDRTCCsConservative.csv')

meta = dat@meta.data
meta = meta %>% rename(XDRTCC = XRTCC)
meta[meta$TCRb_1 %in% xrtcc_CD8$Amino.Acid,]$XDRTCC = 'XDRTCC'
meta[meta$TCRb_1 %in% xrtcc_CD4$Amino.Acid,]$XDRTCC = 'XDRTCC'
meta[meta$TCRb_1 %in% nonxrtcc$Amino.Acid,]$XDRTCC = 'non-XDRTCC'
dat@meta.data = meta
DimPlot_scCustom(dat, split.by = 'POD', group.by = 'XDRTCC', num_columns = 5, colors_use = c('lightgray', 'blue', 'red'), pt.size = 1, order = T) 

# 1.6.25 Edit
cells = meta %>% filter(POD == 'POD33K') %>% filter(XDRTCC = 'XDRTCC')
blood_expanded = c('CASRTGQGVAEAFF', 'CASSDGEVNTEAFF', 'CASSISSYNEQFF', 'CASSVTGTSGRYEQFF', 'CSAREVEPSQETQYF', 'CASSDSTGGTDTQYF', 'CASSENGQPQHF', 'CASSIARRRGAYNEQFF', 'CATSRDQETSGYTF', 'CSASEWGRTEAFF')
intersect(cells$XRTCC_cdr3, blood_expanded)

saveRDS(dat, file = 'integrated6_withoutPigCells.rds')
```

Figure 5B,C
```{r}
markers = FindAllMarkers(dat)
top = markers %>% group_by(cluster) %>% slice_min(n = 25, order_by = p_val_adj)
top = top %>% filter(cluster == '3' | cluster == '5')

sub = subset(dat, idents = c('3', '5'))
sub = ScaleData(sub, features = row.names(sub))
DoHeatmap(sub, features = top$gene, group.colors = c("#CECBD0", "#BD98A2")) + scale_fill_gradientn(colors = c("cornflowerblue","coral3")) +  theme(axis.text.y = element_text(size = 12))  # Adjust size as needed

# This version used for Figures
library(enrichR)
DEenrichRPlot_NS(dat, ident.1 = '3', ident.2 = '5', max.genes = 65, balanced = F, enrich.database = "KEGG_2021_Human")
DEenrichRPlot_NS(dat, ident.1 = '5', ident.2 = '3', max.genes = 65, balanced = F, enrich.database = "KEGG_2021_Human")



markers = FindMarkers(dat, ident.1 = '3', ident.2 = '6')
top = markers %>% slice_max(order_by = avg_log2FC, n = 25)
bottom = markers %>% slice_min(order_by = avg_log2FC, n = 25)

select = c(row.names(top), row.names(bottom))
sub = subset(dat, idents = c('3', '6'))
DoHeatmap(sub, features = select, group.colors = c("#CECBD0", "#BD98A2")) + scale_fill_gradientn(colors =c("cornflowerblue","coral3")) +  theme(text = element_text(size = 20))

markers$gene = row.names(markers)
labels = c('NCAM', 'NCR1', 'NCR3', 'TRDC', 'TRGC1', 'TRGV9', 'TRGV7', 'GZMB', 'GZMK', 'GZMM', 'GZMA', 'GZMH', 'KLRK1', 'CD8B', 'CD3', 'CD3D', 'CD3E', 'CD3G', 'TRBC2')
EnhancedVolcano(markers, lab = row.names(markers), x = 'avg_log2FC', y = 'p_val_adj', drawConnectors = T, pCutoff = 10e-5, max.overlaps = 30, selectLab = labels) + ggtitle('Cluster 3 vs 6')
```

Figure S6
```{r}
# Run Figure 5A code

# Figure S6A
Idents(dat) = 'XDRTCC'
DimPlot_scCustom(dat, split.by = 'XDRTCC', num_columns = 3, colors_use = c('lightgray', 'blue', 'red'), pt.size = 1, order = T, label = F) 

# Figure S6B
table(dat$seurat_clusters2, dat$XDRTCC)

# Figure S6C
# Generate list of Indirect and Direct CD8 CD4
xdrtcc_CD4 = xdrtcc_CD4 %>% mutate(response = ifelse(
  (XKLT001.POD49.MLR.Rcipient.PBMC.vs.dodnor.Pig.PBMC.CD4CFSElow_TCRB > 0 | XKLT001_Direct_LN_v_pig_DC_CD4_CFSE_low_TCRB > 0 |
     XKLT001_POD_28_Direct_MLR_CD4_CFSE_low_TCRB > 0 | XKLT001_Direct_PBMC_v_pig_DC_CD4_CFSE_low_TCRB > 0) &
    XKLT001_Indirect_MLR_CD4_CFSE_low_TCRB == 0, 'CD4 Direct', ifelse(
      (XKLT001.POD49.MLR.Rcipient.PBMC.vs.dodnor.Pig.PBMC.CD4CFSElow_TCRB == 0 | XKLT001_Direct_LN_v_pig_DC_CD4_CFSE_low_TCRB == 0 |
         XKLT001_POD_28_Direct_MLR_CD4_CFSE_low_TCRB == 0 & XKLT001_Direct_PBMC_v_pig_DC_CD4_CFSE_low_TCRB == 0) &
        XKLT001_Indirect_MLR_CD4_CFSE_low_TCRB > 0, 'CD4 Indirect', 'CD4 Undefined')))

xdrtcc_CD8 = xdrtcc_CD8 %>% mutate(response = ifelse(
  (XKLT001.POD49.MLR.Rcipient.PBMC.vs.dodnor.Pig.PBMC.CD8CFSElow_TCRB > 0 | XKLT001_Direct_LN_v_pig_DC_CD8_CFSE_low_TCRB > 0 |
     XKLT001_POD_28_Direct_MLR_CD8_CFSE_low_TCRB > 0 | XKLT001_Direct_PBMC_v_pig_DC_CD8_CFSE_low_TCRB > 0) &
    XKLT001_Indirect_MLR_CD8_CFSE_low_TCRB == 0, 'CD8 Direct', ifelse(
      (XKLT001.POD49.MLR.Rcipient.PBMC.vs.dodnor.Pig.PBMC.CD8CFSElow_TCRB == 0 | XKLT001_Direct_LN_v_pig_DC_CD8_CFSE_low_TCRB == 0 |
         XKLT001_POD_28_Direct_MLR_CD8_CFSE_low_TCRB == 0 & XKLT001_Direct_PBMC_v_pig_DC_CD8_CFSE_low_TCRB == 0) &
        XKLT001_Indirect_MLR_CD8_CFSE_low_TCRB > 0, 'CD8 Indirect', 'CD8 Undefined'))) 

# Determine if XDRTCC in kidney are from indirect or direct
meta = dat@meta.data
meta = meta %>% rownames_to_column('rownmame') %>% left_join(xdrtcc_CD4 %>% select(Amino.Acid, response) %>% rename(response_CD4 = response), by = c("TCRb_1" = "Amino.Acid")) %>% left_join(xdrtcc_CD8 %>% select(Amino.Acid, response) %>% rename(response_CD8 = response), by = c("TCRb_1" = "Amino.Acid")) %>% column_to_rownames('rownmame')
meta = meta %>% mutate(response = case_when(
  is.na(response_CD4) ~ response_CD8, 
  is.na(response_CD8) ~ response_CD4))
meta = meta %>% select(-c(response_CD4, response_CD8))  
meta[is.na(meta$response),]$response = ''
dat@meta.data = meta

DimPlot(dat, split.by = 'POD', group.by = 'response', cols = c('lightgray', "#00BE67", "#C77CFF", "#FDE725FF", "#F8766D", "brown", "#00B4F0") , order = c('CD8 Undefined', 'CD4 Indirect', 'CD8 Indirect', 'CD4 Direct', 'CD8 Direct', ''), pt.size = 1.5)
DimPlot_scCustom(dat, split.by = 'POD', group.by = 'response', colors_use = c('lightgray', "#00BE67", "#C77CFF", "#FDE725FF", "#F8766D", "brown", "#00B4F0") , order = c('CD8 Undefined', 'CD4 Indirect', 'CD8 Indirect', 'CD4 Direct', 'CD8 Direct', ''), pt.size = 1.5, num_columns = 5)
```

Figure 6A
```{r}
Idents(dat) = 'seurat_clusters2'
c3 = WhichCells(dat, idents = '3')
DimPlot(dat, cells.highlight = c3)

sub = readRDS('sub_c3.rds')
meta = sub@meta.data

path = 'Z:/CCTI_USERS/Nathan Suek/2. Collaborations/1. Farshid/2. Decedent xeno/2. Paper Analysis/1. Master Analysis/3. Pig RNAseq Analysis/8. Hybrid Ref v7 - correct GEX original files/'
pig_cells = read.csv(paste0(path,'pig_cells_withPOD61.csv'))
remove = intersect(row.names(meta), pig_cells$pig_cells)

sub = subset(sub, cells = remove, invert = T)
saveRDS(sub, file = 'sub_c3_pigcellremoved.rds')
```

Figure 6B
```{r}
# Define the genes of interest
genes <- c('CD3E', 'TRBC1', 'TRAC', 'TRDC', 'TRGV9', 'NCR1', 'NCR3', 'FCGR3A', 'NCAM1', "GZMB", "GZMA", "GZMK", "GZMH", "PRF1", "GNLY", "KLRD1", "KLRC1", "KLRK1", "ICOS", "PDCD1", "LAG3", "CTLA4", 'TIGIT', 'TCF7') # CD160 and CD244 are not expressed in the XRTCC cluster

# Ensure 'seurat_clusters' is set as the identity class
Idents(sub) <- 'seurat_clusters'

# Get the expression data for the genes of interest
expr_data <- as.matrix(sub@assays$RNA$scale.data[genes, ])

# Extract cluster information for annotation
cluster_info <- as.data.frame(Idents(sub))
colnames(cluster_info) <- "Cluster"

# Extract 'XRTCC' information from metadata
xrtcc_info <- sub@meta.data$XRTCC_cdr3
xrtcc_info[xrtcc_info == ''] <- 'not XDRTCC' 
xrtcc_info <- as.factor(xrtcc_info)
xrtcc_info <- as.data.frame(xrtcc_info) 
colnames(xrtcc_info) <- "XRDTCC CDR3"

# Combine cluster information and 'XRTCC' information
annotation_info <- cbind(cluster_info, xrtcc_info)

# Reorder columns (cells) by cluster
cell_order <- order(annotation_info$Cluster)
expr_data_ordered <- expr_data[, cell_order]
annotation_info_ordered <- annotation_info[cell_order, , drop = FALSE]

# Define the custom color gradient
custom_colors <- colorRampPalette(c("cornflowerblue", "coral3"))(100)

# Generate annotation colors dynamically for clusters
unique_clusters <- unique(annotation_info_ordered$Cluster)
cluster_colors <- setNames(c('#F8766D', '#CD9600', '#7CAE00', '#00C19A', '#00B8E7', '#C77CFF', '#FF61CC'), unique_clusters)

unique_xrtcc <- unique(annotation_info_ordered$`XRDTCC CDR3`)
xrtcc_colors <- setNames(c('white', '#F8766D', '#CD9600', '#7CAE00', '#00C19A', '#00B8E7', '#C77CFF'), unique_xrtcc)

# Combine all annotation colors
anno_colors <- list(Cluster = cluster_colors, 'XRDTCC CDR3' = xrtcc_colors)

# Create the heatmap with the custom color gradient and column annotations
pheatmap(expr_data_ordered,
         annotation_col = annotation_info_ordered,
         cluster_cols = FALSE,
         cluster_rows = FALSE,
         show_rownames = TRUE,
         show_colnames = FALSE,
         main = "Heatmap of Selected Genes by Cluster",
         color = custom_colors,
         annotation_colors = anno_colors)
```

Figure 6E
```{r}

```

Figure s5
T cells: CD3E, CD4, CD8A, CD28
NK cells FCGR3A, NCR1, NCAM1 
APCs: CD80, CD86, HLA-DRB1
```{r}
dat = readRDS('integrated6_withoutPigCells.rds')

meta = dat@meta.data
meta$celltype = ''

# The following definitions are mathematically exclusive
Tcell = WhichCells(dat, expression = CD3E > 0.5 & (TRBC1 > 0.5 | TRDC > 0.5))
NK = WhichCells(dat, expression = CD3E < 0.5 & NCR1 > 0.5 & (FCGR3A > 0.5 | NCAM1 > 0.5))
APC = WhichCells(dat, expression = CD3E < 0.5 & NCR1 < 0.5 & (CD14 > 0.5 | ITGAM > 0.5 | ITGAX > 0.5) & (CD80 > 0.5 | CD86 > 0.5 | 'HLA-DRB1' > 0.5))

meta[Tcell,]$celltype = 'T cell'
meta[NK,]$celltype = 'NK'
meta[APC,]$celltype = 'APC'

dat@meta.data = meta

DimPlot(dat, group.by = 'celltype', cols = c('lightgray', '#00BFC4', '#F8766D', '#7CAE00'), order = c('NK', 'APC', 'T cell', ''), pt.size = 1)

Idents(dat) = 'XDRTCC'
DimPlot_scCustom(dat, split.by = 'XDRTCC', colors_use = c('lightgray', 'blue', 'red'), pt.size = 1, order = T, label = F) 
```

Figure S7
```{r}
DimPlot_scCustom(dat, split.by = 'POD', num_columns = 5, colors_use = scales::hue_pal()(13)) 

highlight = dat@meta.data %>% filter(XRTCC_cdr3 == 'CASSVTGTSGRYEQFF') %>% row.names()
highlight = dat@meta.data %>% filter(XRTCC_cdr3 == 'CASSISSYNEQFF') %>% row.names()
DimPlot_scCustom(dat, split.by = 'POD', num_columns = 5, colors_use = scales::hue_pal()(13)) 

obj.list <- SplitObject(dat, split.by = "POD")
DimPlot(obj.list[[1]], split.by = 'POD', cells.highlight = highlight, pt.size = 2) + NoLegend() + xlim(-10, 20) + ylim(-10, 25)
```

Figure S8
```{r}
# Violint Plots: cluster 3 vs 6
Idents(dat) = 'seurat_clusters2'
sub = subset(dat, idents = (c('3', '5')))

comparison_list <- list(c('3', '5')) # List of comparisons
genes = c('GZMB', 'GZMA', 'GZMK', 'PRF1', 'EOMES', 'PDCD1', 'ICOS', 'TNF', 'IFNG', 'GNLY', 'KLRD1', 'KLRC1')
plots_adjusted <- list() # Initialize a list to store plots

for(i in seq_along(genes)){
  plots <- VlnPlot(sub, features = genes[i]) + NoLegend()
  y_limits <- ggplot_build(plots)$layout$panel_scales_y[[1]]$range$range
  
  # Adjust ylim and add non-overlapping p-value labels
  plots_adjusted[[i]] = plots + 
    stat_compare_means(comparisons = comparison_list, 
                                 label = "p.signif", 
                                 method = 'wilcox.test',
                                 step.increase = 0.2, # Prevents the p-value symbols from overlapping on graph
                                 tip.length = 0) + # Length of the downward tip of each line for each comparison on the graph
          ylim(y_limits[1], y_limits[2] + 0.5) + # Extends y-axis of each graph to accommodate the p-value signs
          theme(axis.title.x = element_blank(), # Hide x-axis title
                axis.title.y = element_blank()) # Hide y-axis title
}
combined_plot <- plot_grid(plotlist = plots_adjusted, ncol = 4, nrow = 3)
combined_plot
```

Figure S9
```{r}
sub = readRDS('sub_c3_pigcellremoved.rds')

AB = WhichCells(sub, expression = TRBC1 > 0.5 & CD3E > 0.5 & TRDC < 0.5, slot = 'data')
DimPlot(sub, cells.highlight = AB)

yd = WhichCells(sub, expression = CD3E > 0.5 & (TRDC > 0.5), slot = 'data')
DimPlot(sub, cells.highlight = yd)

NK = WhichCells(sub, expression = CD3E < 0.5 & (NCAM1 > 0.5 | FCGR3A > 0.5), slot = 'data')
DimPlot(sub, cells.highlight = NK)

# Quantify types of cells per cluster
meta = sub@meta.data
meta$type = ''
meta[AB,]$type = 'AB'
meta[NK,]$type = 'NK'
meta[yd,]$type = 'yd'

sub@meta.data = meta
DimPlot(sub, group.by = 'type')
table(meta$type, Idents(sub))
prop.table(table(meta$type, Idents(sub)), margin = 2)
```

Figure s10
```{r}
# Figure S10A
highlight = WhichCells(dat, expression = CD3E > 0.5 & (TRDC > 0.5), slot = 'data')
obj.list <- SplitObject(dat, split.by = "POD")

lapply(obj.list, function(i){
  DimPlot(i, split.by = 'POD', cells.highlight = highlight, pt.size = 2) + NoLegend() + xlim(-10, 20) + ylim(-10, 25)
})



# Figure S10B
yd_genes = c('TRGV10', 'TRGC1', 'TRGV9', 'TRGV7', 'TRGV5', 'TRGV3', 'TRDC') # when I manually scrolled through the genes, this is all the yd related genes present

# Strategy 1: define yd T cells as above (using GEX definition)
Idents(dat) = "POD"
yd_POD14L = WhichCells(dat, expression = CD3E > 0.5 & (TRDC > 0.5), slot = 'data', ident = 'POD14L')
yd_POD33 = WhichCells(dat, expression = CD3E > 0.5 & (TRDC > 0.5), slot = 'data', ident = 'POD33K')

DimPlot(dat, split.by = 'POD', group.by = 'seurat_clusters2', cells = c(yd_POD14L, yd_POD33))

yd_POD14L = subset(dat, cells = yd_POD14L)
yd_POD33 = subset(dat, cells = yd_POD33)

yd_POD14L_sub = yd_POD14L@assays$RNA$data %>% as.data.frame() %>% filter(row.names(yd_POD14L) %in% yd_genes)
yd_POD33_sub = yd_POD33@assays$RNA$data %>% as.data.frame() %>% filter(row.names(yd_POD33) %in% yd_genes)

yd_POD14L_sub = code_binary(yd_POD14L_sub) # this codes any cell expressing any amount of a gene as 1 (and non expressing as 0)
yd_POD33_sub = code_binary(yd_POD33_sub)

yd_POD14L_sub %>% slice(-c(2:3)) %>% rowSums() %>% prop.table() # TRGV10: 1
yd_POD33_sub %>% slice(-c(2:3)) %>% rowSums() %>% prop.table() # TRGV10: 0 TRGV9: 0.484 TRGV7: 0.303 TRGV5: 0.1212 TRGV3: 0.0909



# Figure S10C
Idents(dat) = "POD"
yd_POD14L = WhichCells(dat, expression = CD3E > 0.5 & (TRDC > 0.5), slot = 'data', ident = 'POD14L')
yd_POD33 = WhichCells(dat, expression = CD3E > 0.5 & (TRDC > 0.5), slot = 'data', ident = 'POD33K')

dat@meta.data$yd = 0
dat@meta.data[yd_POD14L,]$yd = 'POD14L yd'
dat@meta.data[yd_POD33,]$yd = 'POD33 yd'
sub = subset(dat, yd != 0)

comparison_list <- list(c('POD14L', 'POD33K')) # List of comparisons
genes = c('CD3E', 'GZMB', 'GZMA', 'GZMH', 'PRF1', 'GNLY', 'KLRD1', 'KLRC1', 'KLRK1')
plots_adjusted <- list() # Initialize a list to store plots


# Generate the violin plot for the current set of genes
for(i in seq_along(genes)){
  plots <- VlnPlot(sub, features = genes[i]) + NoLegend()
  y_limits <- ggplot_build(plots)$layout$panel_scales_y[[1]]$range$range
  
  # Adjust ylim and add non-overlapping p-value labels
  plots_adjusted[[i]] = plots + 
    stat_compare_means(comparisons = comparison_list, 
                                 label = "p.signif", 
                                 method = 'wilcox.test',
                                 step.increase = 0.2, # Prevents the p-value symbols from overlapping on graph
                                 tip.length = 0) + # Length of the downward tip of each line for each comparison on the graph
          ylim(y_limits[1], y_limits[2] + 0.5) + # Extends y-axis of each graph to accommodate the p-value signs
          theme(axis.title.x = element_blank(), # Hide x-axis title
                axis.title.y = element_blank()) # Hide y-axis title
}
 
# print each plot individually, copy paste into ppt
combined_plot <- plot_grid(plotlist = plots_adjusted, ncol = 5, nrow = 2)
combined_plot
```

Figure s11
```{r}
# Figure S11A
library(celldex)
hpca.se <- HumanPrimaryCellAtlasData()
library(SingleR)

hpca.se <- HumanPrimaryCellAtlasData()
sce <- as.SingleCellExperiment(dat)
pred.main = SingleR(test = sce, assay.type.test = 1, ref = hpca.se, labels = hpca.se$label.main)
pred.fine = SingleR(test = sce, assay.type.test = 1, ref = hpca.se, labels = hpca.se$label.fine)
dat@meta.data$hpca.main = pred.main$pruned.labels
dat@meta.data$hpca.fine = pred.fine$pruned.labels

all.markers <- metadata(pred.main)$de.genes
bcell_markers = all.markers$B_cell %>% unlist() %>% unique()
dc_markers = all.markers %>% unlist() %>% unname() %>% unique()

DimPlot(dat, group.by = 'hpca.main', label = T, repel = T, label.size = 7) + NoLegend() 
DimPlot(dat, group.by = 'hpca.fine', label = T, repel = T) + NoLegend()


# Figure S11B
setwd('Z:/CCTI_SYKES/EXP FF7- XKLT-001 Exp/Bulk TCRb sequincing Adaptive Data XKLT001/All MLRs, Official Analysis')
xrtcc_CD8 = read.csv('MasterCD8DRTCCs_AgainstPreTx.csv')
xrtcc_CD4 = read.csv('MasterCD4DRTCCs_AgainstPreTx.csv')
nonxrtcc = read.csv('PreTxNonDRTCCsConservative.csv')

seurat_obj = dat
meta = seurat_obj@meta.data
tcell = WhichCells(seurat_obj, expression = CD3E > 0.5 & TRBC1 > 0.5 & TRDC < 0.5, slot = 'data')
macrophage_dc = WhichCells(seurat_obj, idents = c(1:6, 8), expression = CD3E < 0.5 & NCR1 < 0.5 & (CD14 > 0.5 | ITGAM > 0.5 | ITGAX > 0.5) & (CD80 > 0.5 | CD86 > 0.5 | 'HLA-DRB1' > 0.5), slot = 'data') # exclude cluster 7 (B cells)
bcell = WhichCells(seurat_obj, idents = '7', slot = 'data')
NK = WhichCells(seurat_obj, expression = CD3E < 0.5 & NCR1 > 0.5 & (FCGR3A > 0.5 | NCAM1 > 0.5), slot = 'data')
yd = WhichCells(seurat_obj, expression = CD3E > 0.5 & TRDC > 0.5, slot = 'data')
DimPlot(seurat_obj, cells.highlight = tcell)
DimPlot(seurat_obj, cells.highlight = macrophage_dc)
DimPlot(seurat_obj, cells.highlight = bcell)
DimPlot(seurat_obj, cells.highlight = NK)
DimPlot(seurat_obj, cells.highlight = yd)

meta$subset_labels = ' '
meta[row.names(meta) %in% tcell,]$subset_labels = 'T cell'
meta[row.names(meta) %in% macrophage_dc,]$subset_labels = 'macrophage_dc'
meta[row.names(meta) %in% bcell,]$subset_labels = 'B cell'
meta[row.names(meta) %in% NK,]$subset_labels = 'NK'
meta[row.names(meta) %in% yd,]$subset_labels = 'yd'
meta[meta$TCRb_1 %in% xrtcc_CD8$Amino.Acid,]$subset_labels = 'XDRTCC'
meta[meta$TCRb_1 %in% xrtcc_CD4$Amino.Acid,]$subset_labels = 'XDRTCC'
meta[meta$TCRb_1 %in% nonxrtcc$Amino.Acid,]$subset_labels = 'non-XDRTCC'
seurat_obj@meta.data = meta

DimPlot_scCustom(seurat_obj, group.by = 'subset_labels', split.by = 'POD', colors_use = c('lightgray', "#F8766D", "#C49A00", "#53B400", "#00C094", 'blue', '#A58AFF', '#FB61D7'), order = c('XDRTCC',  'non-XDRTCC', 'yd', 'NK', 'macrophage_dc', 'B cell', 'T cell', ' '), label = T, repel = T, num_columns = 5)


# Figure S11c
# Cell chat
Idents(seurat_obj) = 'subset_labels'
sub = subset(seurat_obj, idents = c(" "), invert = TRUE)
data.input <- sub[['RNA']]$data
labels <- sub$subset_labels
meta <- data.frame(labels = labels, row.names = names(labels)) # create a dataframe of the cell labels

cellchat <- createCellChat(object = data.input, meta = meta, group.by = "labels")
CellChatDB <- CellChatDB.human  # Use the human ligand-receptor database (or use CellChatDB.mouse for mouse data)
CellChatDB.use <- subsetDB(CellChatDB)
cellchat@DB <- CellChatDB.use
cellchat <- subsetData(cellchat) # This step is necessary even if using the whole database
future::plan("multisession", workers = 4) # do parallel
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)
cellchat <- computeCommunProb(cellchat, type = "triMean")
cellchat <- filterCommunication(cellchat, min.cells = 10)
cellchat <- computeCommunProbPathway(cellchat)
cellchat <- aggregateNet(cellchat)
cellchat@netP$pathways

groupSize <- as.numeric(table(cellchat@idents))
netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")

# Figure 6E
pathways.show <- c("NECTIN") # also use: "IFN-II" "ICAM"
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle")
netVisual_heatmap(cellchat, signaling = pathways.show, color.heatmap = "Reds")
netAnalysis_contribution(cellchat, signaling = pathways.show)
netVisual_bubble(cellchat, signaling = pathways.show, remove.isolate = FALSE, color.heatmap = "viridis") + theme(axis.text.x = element_text(angle = 40, vjust = 1, hjust = 1), text = element_text(size = 20)) 
```

Additional Analysis
Cluster Labeling
```{r}
Idents(dat) = 'seurat_clusters2'
markers = FindAllMarkers(dat)

top = markers %>% group_by(cluster) %>% slice_max(order_by = avg_log2FC, n = 50)
write.csv(top, file = 'cluster.markers.csv')

DimPlot(dat, label = T)
DimPlot_scCustom(dat, split.by = 'POD', num_columns = 5, colors_use = scales::hue_pal()(12)) 

temp = RenameIdents(dat, '0' = 'Naive T cells', '1' = 'Treg and CD4', '2' = 'fibroblasts', '3' = 'NK yd Teff', '4' = 'monocytes macrophages APC', '5' = 'CD8 Teff', '6' = '?human kidney tubule', '7' = 'B cells', '8' = 'NK cells', '9' = '?monocytes', '10'= '?naive T cell', '11' = '? unknown')
DimPlot(temp, label = T, repel = T)
DimPlot_scCustom(temp, split.by = 'POD', num_columns = 5, colors_use = scales::hue_pal()(12)) + NoLegend()
```

Helper Functions
```{r}
# Overview: takes a dataframe and for any cell in the dataframe > 0, it converts it to 1
code_binary = function(df){
  df[df > 0] <- 1
  return(df)
}
```